{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/dask/dataframe/_pyarrow_compat.py:17: FutureWarning: Minimal version of pyarrow will soon be increased to 14.0.1. You are using 13.0.0. Please consider upgrading.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Import required libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import dask.dataframe as dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_Z_prime(pos_mean, pos_std, neg_mean, neg_std):\n",
    "    \"\"\"\n",
    "    This function computes the Z' score for a positive and a negative\n",
    "    control according to the formula provided in the publication\n",
    "    \"Plaque2.0â€”A High-Throughput Analysis Framework to Score Virus-Cell\n",
    "    Transmission and Clonal Cell Expansion\" by Yakimovich et al.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    pos_mean: float\n",
    "        Mean of the positive control.\n",
    "    pos_std: float\n",
    "        Standard deviation of the positive control.\n",
    "    neg_mean: float\n",
    "        Mean of the negative control.\n",
    "    neg_std: float\n",
    "        Standard deviation of the negative control.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Z_prime: float\n",
    "        Z' score computed according to the abovementioned publication.\n",
    "    \"\"\"\n",
    "    Z_prime = (\n",
    "        1\n",
    "        -\n",
    "        (2 * pos_std + 2 * neg_std)\n",
    "        /\n",
    "        abs(pos_mean - neg_mean)\n",
    "    )\n",
    "\n",
    "    return Z_prime\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# As a preliminary step, the Z' scores are computed for the control\n",
    "# wells of the kinome subscreen\n",
    "# Hence, the first step involves loading the respective subscreen\n",
    "# Bear in mind that for certain columns, the data type has to be\n",
    "# manually specified\n",
    "dtype_dict = {\n",
    "    \"Ensembl_ID_OnTarget_Ensembl_GRCh38_release_87\": str,\n",
    "    \"Ensembl_ID_OnTarget_NCBI_HeLa_phs000643_v3_p1_c1_HMB\": str,\n",
    "    \"Gene_Description\": str,\n",
    "    \"ID\": str,\n",
    "    \"ID_OnTarget_Ensembl_GRCh38_release_87\": str,\n",
    "    \"ID_OnTarget_Merge\": str,\n",
    "    \"ID_OnTarget_NCBI_HeLa_phs000643_v3_p1_c1_HMB\": str,\n",
    "    \"ID_OnTarget_RefSeq_20170215\": str,\n",
    "    \"ID_manufacturer\": str,\n",
    "    \"Name_alternatives\": str,\n",
    "    \"PLATE_QUALITY_DESCRIPTION\": str,\n",
    "    \"RefSeq_ID_OnTarget_RefSeq_20170215\": str,\n",
    "    \"Seed_sequence_common\": str,\n",
    "    \"WELL_QUALITY_DESCRIPTION\": str,\n",
    "    \"siRNA_error\": str,\n",
    "    \"siRNA_number\": str,\n",
    "    \"Precursor_Name\": str\n",
    "}\n",
    "\n",
    "main_csv_df = dd.read_csv(\n",
    "    \"VacciniaReport_20170223-0958_ZScored_conc_and_NaN_adjusted.csv\",\n",
    "    sep=\"\\t\",\n",
    "    dtype=dtype_dict\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, extract the subscreen of interest, i.e. the control wells of the\n",
    "# kinome subscreen\n",
    "# It is also important to keep in mind that two different types of\n",
    "# plates have been employed within the framework of the screen, which\n",
    "# are \"CheckerBoard\" and \"ScreeningPlate\"\n",
    "# The difference is that \"CheckerBoard\" plates serve the purpose of\n",
    "# verifying the proper functioning of the hardware, whereas plates of\n",
    "# the type \"ScreeningPlate\" have been used for biological tests in lieu\n",
    "# of hardware tests, i.e. for the actual siRNA-mediated gene knockdowns\n",
    "# To be more precise, plates of the type \"ScreeningPlates\" contain two\n",
    "# types of wells: On the one hand, they harbour control wells, which, as\n",
    "# opposed to the controls in the case of \"CheckerBoard\" plates, do not\n",
    "# deal with the proper functioning of the hardware, but with the\n",
    "# knockdown of target genes the effect of which is known; hence, this\n",
    "# type of control is not about the hardware aspect, but the biology\n",
    "# aspect\n",
    "# On the other hand, plates of the type \"ScreeningPlate\" harbour\n",
    "# treatment wells, i.e. in these wells, genes are knocked down the role\n",
    "# of which in the pathway/process of interest is not known yet\n",
    "# Plates of the type \"ScreeningPlate\" are selected for the purpose of\n",
    "# Z' score computation\n",
    "# Although reliable measurements have not been obtained from all wells,\n",
    "# i.e. there are wells whose value for the feature \"WELL_QUALITY_STATUS\"\n",
    "# is \"BAD\", unreliable measurements are included for now\n",
    "# Note that due to operator precedence, i.e. \"&\" having precedence over\n",
    "# equality checks, the equality checks must be surrounded by equality\n",
    "# checks\n",
    "kinome_control_screening_plates_df = main_csv_df.loc[\n",
    "    (main_csv_df[\"GENESET\"] == \"Kinome\")\n",
    "    &\n",
    "    (main_csv_df[\"WellType\"] == \"CONTROL\")\n",
    "    &\n",
    "    (main_csv_df[\"PLATE_TYPE\"] == \"ScreeningPlate\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ID_list = (\n",
    "    np.unique(kinome_control_screening_plates_df[\"ID_openBIS\"]).tolist()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Questions to ask Artur:\n",
    "1.) The feature \"ID_openBIS\" distinguishes between \"eGFP\" on the one\n",
    "hand and ordinary \"GFP\" on the other hand. I thought that early genes\n",
    "are connected to eGFP and late genes to mCherry as fluorescent markers,\n",
    "so why is ordinary GFP targeted in addition? And what is the difference\n",
    "between eGFP and ordinary GFP?\n",
    "2.) Artur gave two different explanations regarding the control type\n",
    "\"MOCK\": On the one hand, he stated that the MOCK control involved the\n",
    "addition of no viruses such that no intensity is expected. On the other\n",
    "hand, he stated that the MOCK control consisted of infecting cells\n",
    "without siRNA treatment. The mean intensities of early and late genes\n",
    "are comparable to those attained by the SCRAMBLED control, thereby\n",
    "confirming that the latter of the two explanations applies to the MOCK\n",
    "control.\n",
    "3.) Artur told me late gene intensity has been measured with another\n",
    "fluorescence channel; what kind of implications does this have regarding\n",
    "Z' score computation?\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, extract the individual controls comprised in the kinome subscreen\n",
    "# Controls for which their effect is known and hence for which\n",
    "# expectations regarding their intensities exist are the following:\n",
    "# ARAC is the abbreviation for cytosine arabinoside, it is a so-called\n",
    "# antimetabolite which is incorporated into DNA during DNA replication\n",
    "# in lieu of cytidine triphosphate; it inhibits DNA repair, thereby\n",
    "# ultimately resulting in apoptosis; Vaccinia virus does not incorporate\n",
    "# its genome into the host's genome as e.g. retroviruses do and does\n",
    "# therefore not rely on host cell replication for the production and\n",
    "# assembly of progeny constituents; nevertheless, the Wikipedia page\n",
    "# about cytosine arabinoside claimes (without reference) that ARAC is\n",
    "# able to inhibit Vaccinia virus replication; consequently, it might\n",
    "# serve as a (weak) negative control\n",
    "#\n",
    "# ARPC3 is the abbreviation for Actin-related protein 2/3 complex\n",
    "# subunit 3; as its name already suggests, the protein represents one of\n",
    "# seven subunits of the Arp2/3 protein complex; the Vaccinia virus is\n",
    "# known to interact with the actin cytoskeleton and inhibition of the\n",
    "# Arp2/3 complex has been shown to limit vaccinia virus infection;\n",
    "# hence, ARPC3 might serve as a negative control\n",
    "#\n",
    "# ATP6V1A: This gene encodes a subunit of the enzyme vascuolar ATPase,\n",
    "# which effects the acidification of eukaryotic intracellular\n",
    "# organelles; siRNA-mediated knockdown of ATP6V1A prevents virus entry,\n",
    "# which makes insofar sense as both mature virions (MV) and\n",
    "# extracellular virions (EV) of vaccinia viruses rely on the\n",
    "# acidification of the macropinosome for fusion with the macropinosome\n",
    "# membrane to take place\n",
    "# Thus, it is sensible to employ ATP6V1A as negative control, i.e.\n",
    "# control for an unsuccessful host infection\n",
    "#\n",
    "# CDC42 is the abbreviation for cell division control protein 42 homolog\n",
    "# and, as its name already suggests, is involved in the regulation of\n",
    "# the cell cycle; in contrast to other viruses, such as retroviruses,\n",
    "# Vaccinia virus does not incorporate its genome into the host cell's\n",
    "# genome and does therefore not rely on host cell replication for the\n",
    "# poduction of progeny constituents; instead, replication exclusively\n",
    "# takes place in the cytoplasm; hence, while knockdown of CDC42\n",
    "# presumably does not inhibit Vaccinia virus replication altogether, it\n",
    "# is conceivable that the knockdown at least impairs viral replication\n",
    "# as less host cells are available\n",
    "#\n",
    "# CDH4 is the abbreviation for Cadherin-4; hence, it is a so-called\n",
    "# cadherin, i.e. a cell adhesion molecule; cadherins are characterised\n",
    "# by the fact that their activity depends on calcium, hence their name\n",
    "# (the term cadherin is a blend word/portmanteau of \"calcium-dependent\n",
    "# adhesion\"); ...\n",
    "#\n",
    "# CHX is the abbreviation for cyclohexamine; it inhibits translation, or\n",
    "# more precisely translation elongation, in eukaryotic organisms; hence,\n",
    "# even though infection of host cells with Vaccinia virus may be\n",
    "# successful, production and assembly of virus progeny constituents does\n",
    "# not take place or is at least severely impaired; it is therefore\n",
    "# sensible to utilise CHX as negative control\n",
    "#\n",
    "# EGFP is the abbreviation for ...\n",
    "#\n",
    "# FRAP1 is the gene encoding the protein mTOR, i.e. \"mammalian Target of\n",
    "# Rapamycin\"; a brief glance at the scientific literature reveals that\n",
    "# mTOR is intertwined with the vaccinia virus life cycle in a multitude\n",
    "# of ambiguous ways; however, as mTOR is a downstream target of PI3K/Akt\n",
    "# signalling, it is sensible to assume that just as with ITGB1\n",
    "# knockdown, knockdow of mTOR will impede virus entry; this assumption\n",
    "# is corroborated by the mean intensities for both early and late genes,\n",
    "# thereby justifying the usage of FRAP1 as a negative control\n",
    "#\n",
    "# GFP is the abbreviation for ...\n",
    "#\n",
    "# ITGB1 is the abbreviation for integrin beta 1, which is a cell surface\n",
    "# receptor; integrin beta 1 is known to mediate vaccinia virus entry\n",
    "# through activation of PI3K/Akt signalling\n",
    "# (https://journals.asm.org/doi/full/10.1128/jvi.06860-11); hence,\n",
    "# siRNA-mediated knowkdown of ITGB1 impedes vaccinia virus entry and it\n",
    "# is sensible to employ ITGB1 as negative control\n",
    "#\n",
    "# PSMA6 is the abbreviation for Proteasome subunit alpha type-6, which ...\n",
    "#\n",
    "# PXN is the abbreviation for paxillin, the function of which is to\n",
    "# adhere cells to the extracellular matrix; ...\n",
    "#\n",
    "# RAC1 is the abbreviation for Ras-related C3 botulinum toxin substrate\n",
    "# 1, which ...\n",
    "#\n",
    "# SCRAMBLED: These wells involve transfection of cells with\n",
    "# non-targeting siRNAs, i.e. siRNAs that do not effect any gene\n",
    "# knockdown\n",
    "# Consequently, infection of the cells with vaccinia virus is not\n",
    "# impaired, and it is sensible to employ the SCRAMBLED control wells as\n",
    "# positive control, i.e. control for a successful host infection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In order to render the typing process less arduous, the extraction of\n",
    "# the individual controls is not accomplished explicitly, but by looping\n",
    "# through the list containing the IDs\n",
    "individual_control_dfs = []\n",
    "\n",
    "for ID in ID_list:\n",
    "    individual_control_dfs.append(\n",
    "        kinome_control_screening_plates_df.loc[\n",
    "            kinome_control_screening_plates_df[\"ID_openBIS\"] == ID\n",
    "        ]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the intensity values of the entire cell for both early and\n",
    "# late genes\n",
    "# Analogous to the extraction of the individua control data frames, the\n",
    "# extraction of intensity values is not performed explicitly, but by\n",
    "# looping\n",
    "early_int_cells_list = []\n",
    "late_int_cells_list = []\n",
    "\n",
    "for control_df in individual_control_dfs:\n",
    "    early_int_cells_list.append(\n",
    "        control_df[\"dIntensity_cPathogen_eMean_oCells_nZScore\"]\n",
    "    )\n",
    "\n",
    "    late_int_cells_list.append(\n",
    "        control_df[\"dIntensity_cLatePathogen_eMean_oCells_nZScore\"]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, compute for each and every control for both early and late genes\n",
    "# the mean as well as the standard deviation\n",
    "early_mean_list = []\n",
    "early_std_list = []\n",
    "\n",
    "for early_intensity_series in early_int_cells_list:\n",
    "    # Bear in mind that in order to obtain actual values, computation of\n",
    "    # Dask objects has to be triggered\n",
    "    early_mean_list.append(np.mean(early_intensity_series).compute())\n",
    "    early_std_list.append(np.std(early_intensity_series).compute())\n",
    "\n",
    "late_mean_list = []\n",
    "late_std_list = []\n",
    "\n",
    "for late_intensity_series in late_int_cells_list:\n",
    "    late_mean_list.append(np.mean(late_intensity_series).compute())\n",
    "    late_std_list.append(np.std(late_intensity_series).compute())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene:          Early mean:   Late mean:\n",
      "---------------------------------------\n",
      "ABI1           1.051         0.674\n",
      "---------------------------------------\n",
      "ALLSTARDEATH   -1.241        4.323\n",
      "---------------------------------------\n",
      "ARPC3          -0.502        -0.236\n",
      "---------------------------------------\n",
      "ATP6V1A        0.233         0.499\n",
      "---------------------------------------\n",
      "CDC42          0.714         1.388\n",
      "---------------------------------------\n",
      "CDH4           0.71          0.229\n",
      "---------------------------------------\n",
      "EGFP           -2.806        -0.2\n",
      "---------------------------------------\n",
      "FRAP1          -1.113        -0.29\n",
      "---------------------------------------\n",
      "GFP            -2.641        0.007\n",
      "---------------------------------------\n",
      "ITGB1          -0.784        -0.152\n",
      "---------------------------------------\n",
      "KIF11          0.504         3.049\n",
      "---------------------------------------\n",
      "MAP3K7         0.904         0.966\n",
      "---------------------------------------\n",
      "MET            0.839         0.264\n",
      "---------------------------------------\n",
      "MOCK           0.522         0.462\n",
      "---------------------------------------\n",
      "PAK1           0.646         0.76\n",
      "---------------------------------------\n",
      "PI4KB          0.221         0.018\n",
      "---------------------------------------\n",
      "PSMA6          0.778         1.024\n",
      "---------------------------------------\n",
      "PSMC3          0.149         0.438\n",
      "---------------------------------------\n",
      "PXN            1.094         1.092\n",
      "---------------------------------------\n",
      "RAC1           0.567         1.207\n",
      "---------------------------------------\n",
      "SCRAMBLED      0.472         0.77\n",
      "---------------------------------------\n",
      "SISEL_NC1      -0.476        0.138\n",
      "---------------------------------------\n",
      "SISEL_NC2      0.275         0.132\n",
      "---------------------------------------\n",
      "TSG101         1.088         1.135\n",
      "---------------------------------------\n",
      "UNKNOWN        0.781         0.546\n"
     ]
    }
   ],
   "source": [
    "# The mean intensities for early and late genes are visually inspected\n",
    "# for all controls\n",
    "# Determine the longest gene name in order to appropriately set the\n",
    "# width of the gene name column\n",
    "max_gene_name_len = max([len(name) for name in ID_list])\n",
    "second_column_name = \"Early mean:\"\n",
    "second_column_len = len(second_column_name)\n",
    "third_column_name = \"Late mean:\"\n",
    "third_column_len = len(third_column_name)\n",
    "\n",
    "# Bear in mind that the default value of the \"sep\" argument of the print\n",
    "# function is used, which is a single space character\n",
    "# Hence, in order to determine the total amount of characters one row\n",
    "# encompasses, two characters have to be added in order to take into\n",
    "# account the transition from the first to the second column and the\n",
    "# transition from the second to the third column, respectively\n",
    "# Also keep in mind the two additional spaces incorporated between\n",
    "# columns\n",
    "total_row_len = (\n",
    "    max_gene_name_len + 2 + 1 + second_column_len + 2 + 1 + third_column_len\n",
    ")\n",
    "\n",
    "print(\n",
    "    \"Gene:\".ljust(max_gene_name_len + 2),\n",
    "    second_column_name.ljust(second_column_len + 2),\n",
    "    third_column_name\n",
    ")\n",
    "\n",
    "for i, control_name in enumerate(ID_list):\n",
    "    current_control_early_mean = early_mean_list[i]\n",
    "    current_control_late_mean = late_mean_list[i]\n",
    "    print(\"-\" * total_row_len)\n",
    "    print(\n",
    "        control_name.ljust(max_gene_name_len + 2),\n",
    "        str(np.round(\n",
    "            current_control_early_mean, 3\n",
    "        )).ljust(second_column_len + 2),\n",
    "        np.round(current_control_late_mean, 3)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Z' scores for the following combinations of positive and negative controls:\n",
      "SCRAMBLED-ARPC3 early: -2.191\n",
      "SCRAMBLED-ARPC3 late: -2.586\n",
      "MOCK-ARPC3 early: -2.257\n",
      "MOCK-ARPC3 late: -5.052\n",
      "\n",
      "Z' scores for the following combinations of positive and negative controls:\n",
      "SCRAMBLED-ATP6V1A early: -11.219\n",
      "SCRAMBLED-ATP6V1A late: -12.981\n",
      "MOCK-ATP6V1A early: -9.906\n",
      "MOCK-ATP6V1A late: -112.396\n",
      "\n",
      "Z' scores for the following combinations of positive and negative controls:\n",
      "SCRAMBLED-FRAP1 early: -0.356\n",
      "SCRAMBLED-FRAP1 late: -1.806\n",
      "MOCK-FRAP1 early: -0.452\n",
      "MOCK-FRAP1 late: -3.574\n",
      "\n",
      "Z' scores for the following combinations of positive and negative controls:\n",
      "SCRAMBLED-ITGB1 early: -0.895\n",
      "SCRAMBLED-ITGB1 late: -2.304\n",
      "MOCK-ITGB1 early: -0.995\n",
      "MOCK-ITGB1 late: -4.718\n"
     ]
    }
   ],
   "source": [
    "# Now, a couple of Z' scores are computed\n",
    "# The only two positive controls identified thus far are SCRAMBLED and\n",
    "# MOCK\n",
    "SCRAMBLED_index = ID_list.index(\"SCRAMBLED\")\n",
    "SCRAMBLED_early_mean = early_mean_list[SCRAMBLED_index]\n",
    "SCRAMBLED_early_std = early_std_list[SCRAMBLED_index]\n",
    "SCRAMBLED_late_mean = late_mean_list[SCRAMBLED_index]\n",
    "SCRAMBLED_late_std = late_std_list[SCRAMBLED_index]\n",
    "\n",
    "MOCK_index = ID_list.index(\"MOCK\")\n",
    "MOCK_early_mean = early_mean_list[MOCK_index]\n",
    "MOCK_early_std = early_std_list[MOCK_index]\n",
    "MOCK_late_mean = late_mean_list[MOCK_index]\n",
    "MOCK_late_std = late_std_list[MOCK_index]\n",
    "\n",
    "# Z' score for ARPC3 as negative control\n",
    "ARPC3_index = ID_list.index(\"ARPC3\")\n",
    "ARPC3_early_mean = early_mean_list[ARPC3_index]\n",
    "ARPC3_early_std = early_std_list[ARPC3_index]\n",
    "ARPC3_late_mean = late_mean_list[ARPC3_index]\n",
    "ARPC3_late_std = late_std_list[ARPC3_index]\n",
    "\n",
    "Z_prime_early_ARPC3_SCRAMBLED = compute_Z_prime(\n",
    "    SCRAMBLED_early_mean, SCRAMBLED_early_std,\n",
    "    ARPC3_early_mean, ARPC3_early_std\n",
    ")\n",
    "Z_prime_late_ARPC3_SCRAMBLED = compute_Z_prime(\n",
    "    SCRAMBLED_late_mean, SCRAMBLED_late_std,\n",
    "    ARPC3_late_mean, SCRAMBLED_late_std\n",
    ")\n",
    "\n",
    "Z_prime_early_ARPC3_MOCK = compute_Z_prime(\n",
    "    MOCK_early_mean, MOCK_early_std,\n",
    "    ARPC3_early_mean, ARPC3_early_std\n",
    ")\n",
    "Z_prime_late_ARPC3_MOCK = compute_Z_prime(\n",
    "    MOCK_late_mean, MOCK_late_std,\n",
    "    ARPC3_late_mean, ARPC3_late_std\n",
    ")\n",
    "\n",
    "print(\n",
    "    \"Z' scores for the following combinations of positive and \"\n",
    "    \"negative controls:\\n\"\n",
    "    f\"SCRAMBLED-ARPC3 early: {np.round(Z_prime_early_ARPC3_SCRAMBLED, 3)}\\n\"\n",
    "    f\"SCRAMBLED-ARPC3 late: {np.round(Z_prime_late_ARPC3_SCRAMBLED, 3)}\\n\"\n",
    "    f\"MOCK-ARPC3 early: {np.round(Z_prime_early_ARPC3_MOCK, 3)}\\n\"\n",
    "    f\"MOCK-ARPC3 late: {np.round(Z_prime_late_ARPC3_MOCK, 3)}\"\n",
    ")\n",
    "print()\n",
    "\n",
    "# Z' score for ATP6V1A as negative control\n",
    "ATP6V1A_index = ID_list.index(\"ATP6V1A\")\n",
    "ATP6V1A_early_mean = early_mean_list[ATP6V1A_index]\n",
    "ATP6V1A_early_std = early_std_list[ATP6V1A_index]\n",
    "ATP6V1A_late_mean = late_mean_list[ATP6V1A_index]\n",
    "ATP6V1A_late_std = late_std_list[ATP6V1A_index]\n",
    "\n",
    "Z_prime_early_ATP6V1A_SCRAMBLED = compute_Z_prime(\n",
    "    SCRAMBLED_early_mean, SCRAMBLED_early_std,\n",
    "    ATP6V1A_early_mean, ATP6V1A_early_std\n",
    ")\n",
    "Z_prime_late_ATP6V1A_SCRAMBLED = compute_Z_prime(\n",
    "    SCRAMBLED_late_mean, SCRAMBLED_late_std,\n",
    "    ATP6V1A_late_mean, ATP6V1A_late_std\n",
    ")\n",
    "\n",
    "Z_prime_early_ATP6V1A_MOCK = compute_Z_prime(\n",
    "    MOCK_early_mean, MOCK_early_std,\n",
    "    ATP6V1A_early_mean, ATP6V1A_early_std\n",
    ")\n",
    "Z_prime_late_ATP6V1A_MOCK = compute_Z_prime(\n",
    "    MOCK_late_mean, MOCK_late_std,\n",
    "    ATP6V1A_late_mean, ATP6V1A_late_std\n",
    ")\n",
    "\n",
    "print(\n",
    "    \"Z' scores for the following combinations of positive and \"\n",
    "    \"negative controls:\\n\"\n",
    "    f\"SCRAMBLED-ATP6V1A early: {np.round(Z_prime_early_ATP6V1A_SCRAMBLED, 3)}\\n\"\n",
    "    f\"SCRAMBLED-ATP6V1A late: {np.round(Z_prime_late_ATP6V1A_SCRAMBLED, 3)}\\n\"\n",
    "    f\"MOCK-ATP6V1A early: {np.round(Z_prime_early_ATP6V1A_MOCK, 3)}\\n\"\n",
    "    f\"MOCK-ATP6V1A late: {np.round(Z_prime_late_ATP6V1A_MOCK, 3)}\"\n",
    ")\n",
    "print()\n",
    "\n",
    "# Z' score for FRAP1 as negative control\n",
    "FRAP1_index = ID_list.index(\"FRAP1\")\n",
    "FRAP1_early_mean = early_mean_list[FRAP1_index]\n",
    "FRAP1_early_std = early_std_list[FRAP1_index]\n",
    "FRAP1_late_mean = late_mean_list[FRAP1_index]\n",
    "FRAP1_late_std = late_std_list[FRAP1_index]\n",
    "\n",
    "Z_prime_early_FRAP1_SCRAMBLED = compute_Z_prime(\n",
    "    SCRAMBLED_early_mean, SCRAMBLED_early_std,\n",
    "    FRAP1_early_mean, FRAP1_early_std\n",
    ")\n",
    "Z_prime_late_FRAP1_SCRAMBLED = compute_Z_prime(\n",
    "    SCRAMBLED_late_mean, SCRAMBLED_late_std,\n",
    "    FRAP1_late_mean, FRAP1_late_std\n",
    ")\n",
    "\n",
    "Z_prime_early_FRAP1_MOCK = compute_Z_prime(\n",
    "    MOCK_early_mean, MOCK_early_std,\n",
    "    FRAP1_early_mean, FRAP1_early_std\n",
    ")\n",
    "Z_prime_late_FRAP1_MOCK = compute_Z_prime(\n",
    "    MOCK_late_mean, MOCK_late_std,\n",
    "    FRAP1_late_mean, FRAP1_late_std\n",
    ")\n",
    "\n",
    "print(\n",
    "    \"Z' scores for the following combinations of positive and \"\n",
    "    \"negative controls:\\n\"\n",
    "    f\"SCRAMBLED-FRAP1 early: {np.round(Z_prime_early_FRAP1_SCRAMBLED, 3)}\\n\"\n",
    "    f\"SCRAMBLED-FRAP1 late: {np.round(Z_prime_late_FRAP1_SCRAMBLED, 3)}\\n\"\n",
    "    f\"MOCK-FRAP1 early: {np.round(Z_prime_early_FRAP1_MOCK, 3)}\\n\"\n",
    "    f\"MOCK-FRAP1 late: {np.round(Z_prime_late_FRAP1_MOCK, 3)}\"\n",
    ")\n",
    "print()\n",
    "\n",
    "# Z' score for ITGB1 as negative control\n",
    "ITGB1_index = ID_list.index(\"ITGB1\")\n",
    "ITGB1_early_mean = early_mean_list[ITGB1_index]\n",
    "ITGB1_early_std = early_std_list[ITGB1_index]\n",
    "ITGB1_late_mean = late_mean_list[ITGB1_index]\n",
    "ITGB1_late_std = late_std_list[ITGB1_index]\n",
    "\n",
    "Z_prime_early_ITGB1_SCRAMBLED = compute_Z_prime(\n",
    "    SCRAMBLED_early_mean, SCRAMBLED_early_std,\n",
    "    ITGB1_early_mean, ITGB1_early_std\n",
    ")\n",
    "Z_prime_late_ITGB1_SCRAMBLED = compute_Z_prime(\n",
    "    SCRAMBLED_late_mean, SCRAMBLED_late_std,\n",
    "    ITGB1_late_mean, ITGB1_late_std\n",
    ")\n",
    "\n",
    "Z_prime_early_ITGB1_MOCK = compute_Z_prime(\n",
    "    MOCK_early_mean, MOCK_early_std,\n",
    "    ITGB1_early_mean, ITGB1_early_std\n",
    ")\n",
    "Z_prime_late_ITGB1_MOCK = compute_Z_prime(\n",
    "    MOCK_late_mean, MOCK_late_std,\n",
    "    ITGB1_late_mean, ITGB1_late_std\n",
    ")\n",
    "\n",
    "print(\n",
    "    \"Z' scores for the following combinations of positive and \"\n",
    "    \"negative controls:\\n\"\n",
    "    f\"SCRAMBLED-ITGB1 early: {np.round(Z_prime_early_ITGB1_SCRAMBLED, 3)}\\n\"\n",
    "    f\"SCRAMBLED-ITGB1 late: {np.round(Z_prime_late_ITGB1_SCRAMBLED, 3)}\\n\"\n",
    "    f\"MOCK-ITGB1 early: {np.round(Z_prime_early_ITGB1_MOCK, 3)}\\n\"\n",
    "    f\"MOCK-ITGB1 late: {np.round(Z_prime_late_ITGB1_MOCK, 3)}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the CSV file\n",
    "train_test_df = pd.read_csv(\"180828_train_test.csv\")\n",
    "\n",
    "# Retrieve rows bearing the name \"SCRAMBLED\", i.e. representing siRNA\n",
    "# having no effect -> negative control\n",
    "# Note that in the context of the manuscript, this data type is referred\n",
    "# to as \"average\"\n",
    "average_df = train_test_df.loc[train_test_df[\"Name\"] == \"SCRAMBLED\"]\n",
    "\n",
    "# Retrieve rows bearing the name \"GFP Duplex III\", i.e. representing the\n",
    "# targeted knock-down of the GFP gene\n",
    "# Consequently, \"GFP Duplex III\" represents the positive control\n",
    "gfp_duplex_df = train_test_df.loc[train_test_df[\"Name\"] == \"GFP Duplex III\"]\n",
    "\n",
    "# Likewise, retrieve columns for the remaining categories, which are\n",
    "# early and late genes low, early genes high and late genes low, early\n",
    "# genes low and late genes high, early and late genes high, cell death\n",
    "# and mock\n",
    "class_2_df = train_test_df.loc[train_test_df[\"Class\"] == 2]\n",
    "class_3_df = train_test_df.loc[train_test_df[\"Class\"] == 3]\n",
    "class_4_df = train_test_df.loc[train_test_df[\"Class\"] == 4]\n",
    "# Note that due to operator precedence, i.e. \"&\" having precedence over\n",
    "# equality checks, the equality checks must be surrounded by parentheses\n",
    "class_5_df = train_test_df.loc[\n",
    "    (train_test_df[\"Class\"] == 5) & (train_test_df[\"Name\"] != \"Mock\")\n",
    "]\n",
    "cell_death_df = train_test_df.loc[train_test_df[\"Class\"] == 6]\n",
    "mock_df = train_test_df.loc[train_test_df[\"Name\"] == \"Mock\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Positive control (GFP Duplex) early mean and std: (0.07858209514814343, 0.024904375138366207)\n",
      "Negative control (SCRAMBLED) early mean and std: (0.2693598656935352, 0.04430401822853731)\n",
      "\n",
      "Positive control (GFP Duplex) late mean and std: (0.05958844995024221, 0.01686860452587924)\n",
      "Negative control (SCRAMBLED) late mean and std: (0.06502387037258052, 0.014771989411226984)\n",
      "0.274460612795171\n",
      "-10.642372246706357\n"
     ]
    }
   ],
   "source": [
    "# Compute the Z' score exemplarily for \"GFP Duplex III\" as positive\n",
    "# control and \"SCRAMBLED\" as negative control\n",
    "# As a preliminary step, the intensity values of the peri-nucleus are\n",
    "# extracted for both early and late genes\n",
    "# Additionally, for the sake of convenience, the extracted data series\n",
    "# are converted into NumPy arrays\n",
    "gfp_duplex_early_int_perinuc = gfp_duplex_df[\"dIntensity_cPathogen_eMean_oPeriNuclei_nZScore\"].to_numpy()\n",
    "gfp_duplex_late_int_perinuc = gfp_duplex_df[\"dIntensity_cLatePathogen_eMean_oPeriNuclei_nZScore\"].to_numpy()\n",
    "\n",
    "average_early_int_perinuc = average_df[\"dIntensity_cPathogen_eMean_oPeriNuclei_nZScore\"].to_numpy()\n",
    "average_late_int_perinuc = average_df[\"dIntensity_cLatePathogen_eMean_oPeriNuclei_nZScore\"].to_numpy()\n",
    "\n",
    "gfp_duplex_early_mean = np.mean(gfp_duplex_early_int_perinuc)\n",
    "gfp_duplex_early_std = np.std(gfp_duplex_early_int_perinuc)\n",
    "\n",
    "gfp_duplex_late_mean = np.mean(gfp_duplex_late_int_perinuc)\n",
    "gfp_duplex_late_std = np.std(gfp_duplex_late_int_perinuc)\n",
    "\n",
    "average_early_mean = np.mean(average_early_int_perinuc)\n",
    "average_early_std = np.std(average_early_int_perinuc)\n",
    "\n",
    "average_late_mean = np.mean(average_late_int_perinuc)\n",
    "average_late_std = np.std(average_late_int_perinuc)\n",
    "\n",
    "print(f\"Positive control (GFP Duplex) early mean and std: {gfp_duplex_early_mean, gfp_duplex_early_std}\")\n",
    "print(f\"Negative control (SCRAMBLED) early mean and std: {average_early_mean, average_early_std}\")\n",
    "print()\n",
    "print(f\"Positive control (GFP Duplex) late mean and std: {gfp_duplex_late_mean, gfp_duplex_late_std}\")\n",
    "print(f\"Negative control (SCRAMBLED) late mean and std: {average_late_mean, average_late_std}\")\n",
    "\n",
    "Z_prime_early = compute_Z_prime(\n",
    "    gfp_duplex_early_mean,\n",
    "    gfp_duplex_early_std,\n",
    "    average_early_mean,\n",
    "    average_early_std\n",
    ")\n",
    "Z_prime_late = compute_Z_prime(\n",
    "    gfp_duplex_late_mean,\n",
    "    gfp_duplex_late_std,\n",
    "    average_late_mean,\n",
    "    average_late_std\n",
    ")\n",
    "\n",
    "print(Z_prime_early)\n",
    "print(Z_prime_late)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each category, extract the intensity values of the entire cell for\n",
    "# both early and late genes\n",
    "# For the sake of convenience, the extracted data series are converted\n",
    "# into NumPy arrays\n",
    "average_early_int_cell = average_df[\"dIntensity_cPathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "average_late_int_cell = average_df[\"dIntensity_cLatePathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "\n",
    "class_2_early_int_cell = class_2_df[\"dIntensity_cPathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "class_2_late_int_cell = class_2_df[\"dIntensity_cLatePathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "\n",
    "class_3_early_int_cell = class_3_df[\"dIntensity_cPathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "class_3_late_int_cell = class_3_df[\"dIntensity_cLatePathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "\n",
    "class_4_early_int_cell = class_4_df[\"dIntensity_cPathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "class_4_late_int_cell = class_4_df[\"dIntensity_cLatePathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "\n",
    "class_5_early_int_cell = class_5_df[\"dIntensity_cPathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "class_5_late_int_cell = class_5_df[\"dIntensity_cLatePathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "\n",
    "cell_death_early_int_cell = cell_death_df[\"dIntensity_cPathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "cell_death_late_int_cell = cell_death_df[\"dIntensity_cLatePathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "\n",
    "mock_early_int_cell = mock_df[\"dIntensity_cPathogen_eMean_oCells_nZScore\"].to_numpy()\n",
    "mock_late_int_cell = mock_df[\"dIntensity_cLatePathogen_eMean_oCells_nZScore\"].to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Z' score class 2 early genes: -6.558581961073892\n",
      "Z' score class 2 late genes: -12.057857737179544\n"
     ]
    }
   ],
   "source": [
    "# Compute Z' score exemplarily for class 2 with average as negative\n",
    "# control\n",
    "class_2_early_mean = np.mean(class_2_early_int_cell)\n",
    "class_2_early_std = np.std(class_2_early_int_cell)\n",
    "\n",
    "class_2_late_mean = np.mean(class_2_late_int_cell)\n",
    "class_2_late_std = np.std(class_2_late_int_cell)\n",
    "\n",
    "average_early_mean = np.mean(average_early_int_cell)\n",
    "average_early_std = np.std(average_early_int_cell)\n",
    "\n",
    "average_late_mean = np.mean(average_late_int_cell)\n",
    "average_late_std = np.std(average_late_int_cell)\n",
    "\n",
    "Z_prime_early = (\n",
    "    1\n",
    "    -\n",
    "    (3 * class_2_early_std + 3 * average_early_std)\n",
    "    /\n",
    "    abs(class_2_early_mean - average_early_mean)\n",
    ")\n",
    "Z_prime_late = (\n",
    "    1\n",
    "    -\n",
    "    (3 * class_2_late_std + 3 * average_late_std)\n",
    "    /\n",
    "    abs(class_2_late_mean - average_late_mean)\n",
    ")\n",
    "\n",
    "print(f\"Z' score class 2 early genes: {Z_prime_early}\")\n",
    "print(f\"Z' score class 2 late genes: {Z_prime_late}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinformatics",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
