{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from biotite.database import entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unfortunately, to my knowledge, downloading all VACV interactions from\n",
    "# the HVIDB database at once is not possible\n",
    "# Hence, they have been downloaded in packages of up to 100 interactions\n",
    "# It is investigated whether the 5 packages are all different from each\n",
    "# other so as to rule out errors during the manual download, such as\n",
    "# accidentally downloading the same package twice\n",
    "# As a first step, the individual packages are loaded into the Jupyter\n",
    "# notebook\n",
    "HVIDB_package_1 = pd.read_csv(\"HVIDB_VACV_interactions_1.csv\")\n",
    "HVIDB_package_2 = pd.read_csv(\"HVIDB_VACV_interactions_2.csv\")\n",
    "HVIDB_package_3 = pd.read_csv(\"HVIDB_VACV_interactions_3.csv\")\n",
    "HVIDB_package_4 = pd.read_csv(\"HVIDB_VACV_interactions_4.csv\")\n",
    "HVIDB_package_5 = pd.read_csv(\"HVIDB_VACV_interactions_5.csv\")\n",
    "\n",
    "# According to the HVIDB database, the total amount of interactions is\n",
    "# 456\n",
    "# It is checked whether this also applies to the five downloaded\n",
    "# packages\n",
    "packages_list = [\n",
    "    HVIDB_package_1,\n",
    "    HVIDB_package_2,\n",
    "    HVIDB_package_3,\n",
    "    HVIDB_package_4,\n",
    "    HVIDB_package_5\n",
    "]\n",
    "\n",
    "assert sum(map(len, packages_list)) == 456, (\n",
    "    \"An error has been made during the manual download of the \"\n",
    "    \"interaction packages!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, verify that all downloaded interaction packages are distinct from\n",
    "# one another\n",
    "# Unfortunately, checking equality for more than two DataFrames at once\n",
    "# is not possible to the best of my knowledge\n",
    "# Hence, it is resorted to for-loops\n",
    "package_equality = False\n",
    "\n",
    "for i, package in enumerate(packages_list):\n",
    "    # For obvious reasons, equality checks are only performed between\n",
    "    # different interaction packages\n",
    "    # Moreover, redundant equality checks are avoided\n",
    "    for j in range(i + 1, 5):\n",
    "        package_equality = package_equality or package.equals(packages_list[j])\n",
    "\n",
    "assert package_equality == False, (\n",
    "    \"Identical interaction packages have erroneously been downloaded!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, for the sake of convenience, concatenate the five individual\n",
    "# interaction packages to generate one coherent CSV file\n",
    "HVIDB_VACV_interactions_df = pd.concat(packages_list, ignore_index=True)\n",
    "\n",
    "assert (\n",
    "    len(HVIDB_VACV_interactions_df) == 456\n",
    "    and\n",
    "    len(HVIDB_VACV_interactions_df.columns) == 4\n",
    "), \"A mistake was done during DataFrame concatenation!\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the coherent DataFrame as a CSV file\n",
    "HVIDB_VACV_interactions_df.to_csv(\n",
    "    \"all_HVIDB_VACV_interactions.csv\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amount of human proteins targeted in the VACV screen by single/pooled siRNAs as well as esiRNAs: 18,395\n"
     ]
    }
   ],
   "source": [
    "# Determine the amount of genes targeted by siRNAs/esiRNAs and in the\n",
    "# VACV screen\n",
    "# Low-molecular weight organic compounds are taken into account as\n",
    "# despite having the same final outcome as siRNAs, i.e. the inhibition\n",
    "# of the action of proteins, low-molecular weight compounds and siRNAs\n",
    "# differ from one another in their mode of action: The former come into\n",
    "# play after protein translation has taken place, whereas the latter\n",
    "# unfold their effect on the post-transcriptional level by targeting\n",
    "# mRNAs\n",
    "# Hence, exactly the same effect is not expected\n",
    "single_pooled_siRNA_and_esiRNA_STRING_IDs_df = pd.read_csv(\n",
    "    \"single_pooled_siRNA_and_esiRNA_STRING_IDs.csv\",\n",
    "    delimiter=\"\\t\"\n",
    ")\n",
    "\n",
    "n_targets = len(single_pooled_siRNA_and_esiRNA_STRING_IDs_df)\n",
    "\n",
    "print(\n",
    "    \"Amount of human proteins targeted in the VACV screen by \"\n",
    "    f\"single/pooled siRNAs as well as esiRNAs: {n_targets:,}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, determine the amount of human genes interrogated in the VACV\n",
    "# screen that engage in PPIs recorded by HVIDB\n",
    "# To this end, the protein target names have to be converted in\n",
    "# UniProtIDs as they are employed by HVIDB\n",
    "# This is achieved via the ID mapping GUI on\n",
    "# https://www.uniprot.org/id-mapping\n",
    "VACV_screen_human_targets_series = (\n",
    "    single_pooled_siRNA_and_esiRNA_STRING_IDs_df[\"queryItem\"]\n",
    ")\n",
    "\n",
    "VACV_screen_human_targets_series.to_csv(\n",
    "    \"VACV_screen_human_targets_single_pooled_siRNA_and_esiRNA_gene_names.txt\",\n",
    "    header=False,\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using the parameters \"from: Gene_Name\", \"to: UniProtKB-Swiss-Prot\" and\n",
    "# \"taxId: 9606\", 18,185 of the 18,396 human protein IDs were mapped to\n",
    "# 18,645 results, whereas 213 ID were not mapped\n",
    "# This is subjected to close scrutiny\n",
    "#\n",
    "# For the ALG1L gene, the official gene symbol has apparently been\n",
    "# changed after the VACV screen has been conducted; its new official\n",
    "# gene symbol is ALG1L1P while its NCBI Gene ID remained unchanged\n",
    "# (200810)\n",
    "#\n",
    "# Regarding the gene AKAP2, the NCBI Gene ID has been changed from 11217\n",
    "# to 445815; note that AKAP2 is only an alias and that the offical gene\n",
    "# symbol is PALM2AKAP2\n",
    "#\n",
    "# With respect to the gene PPAN-P2RY11, neither the official gene symbol\n",
    "# nor the NCBI Gene ID have been altered after the VACV screen, but the\n",
    "# respective UniProtKB entry belongs to the UniProtKB/TrEMBL section,\n",
    "# i.e. has not been manually reviewed. Consequently, using the setting\n",
    "# \"to: UniProtKB-Swiss-Prot\" does not yield any UniProtID for this gene\n",
    "# It becomes apparent that the most convenient way of obtaining\n",
    "# UniProtIDs in one single step is to use the settings\n",
    "# \"From database: Genomic annotation databases/GeneID\" (Gene ID is the\n",
    "# same as NCBI Gene ID) and \"To database: UniProt/UniProtKB\" (choosing\n",
    "# UniProtKB includes both UniProtKB sections, i.e. UniProtKB/Swiss-Prot\n",
    "# and UniProtKB/TrEMBL)\n",
    "#\n",
    "# With regard to the gene CDY, the specified NCBI Gene ID is still valid\n",
    "# (203611), but its official gene symbol is CDY2B\n",
    "#\n",
    "# As to the gene FLJ40296, the specified NCBI Gene ID is still valid\n",
    "# (122183), but its official gene symbol is PRR20A\n",
    "#\n",
    "# Regarding the gene MGC39606, the specified NCBI Gene ID is still valid\n",
    "# (399668), but its official gene symbol is SMIM10L2A\n",
    "#\n",
    "# etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Instead of doing this investigation manually, it is accomplished in an\n",
    "# automated fashion by means of the Biotite bioinformatics library\n",
    "# Biotite provides functionalities allowing to access various NCBI\n",
    "# databases in a programmatic manner, i.e. from within Python scripts\n",
    "\n",
    "# Two columns of interest for the following endeavour are \"ID\" as well\n",
    "# as \"ID_manufacturer\"\n",
    "# Upon closer scrutiny, it became apparent that contrary to their names,\n",
    "# both features basically harbour the same information, namely the NCBI\n",
    "# Gene ID of the targeted gene\n",
    "# However, they differ in that while the feature \"ID\" is populated\n",
    "# depending on whether the respective experiment was successful or not,\n",
    "# \"ID_manufacturer\" is continuously populated, irrespective of the\n",
    "# experiment's outcome\n",
    "# Hence, the entries of the \"ID_manufacturer\" feature are employed in\n",
    "# order to query NCBI's gene database\n",
    "# As a first step, it is checker whether all entries of the\n",
    "# \"ID_manufacturer\" feature are indeed populated with values or whether\n",
    "# there are isolated occurrences of \"Not available\"\n",
    "# Note that the features \"ID\" and \"ID_manufacturer\" only harbour NCBI\n",
    "# Gene IDs in the case of siRNA, pooled siRNA and esiRNA\n",
    "\n",
    "# Load the screen data\n",
    "# Bear in mind that for certain columns, the data type has to be\n",
    "# manually specified\n",
    "dtype_dict = {\n",
    "    \"Ensembl_ID_OnTarget_Ensembl_GRCh38_release_87\": str,\n",
    "    \"Ensembl_ID_OnTarget_NCBI_HeLa_phs000643_v3_p1_c1_HMB\": str,\n",
    "    \"Gene_Description\": str,\n",
    "    \"ID\": str,\n",
    "    \"ID_OnTarget_Ensembl_GRCh38_release_87\": str,\n",
    "    \"ID_OnTarget_Merge\": str,\n",
    "    \"ID_OnTarget_NCBI_HeLa_phs000643_v3_p1_c1_HMB\": str,\n",
    "    \"ID_OnTarget_RefSeq_20170215\": str,\n",
    "    \"ID_manufacturer\": str,\n",
    "    \"Name_alternatives\": str,\n",
    "    \"PLATE_QUALITY_DESCRIPTION\": str,\n",
    "    \"RefSeq_ID_OnTarget_RefSeq_20170215\": str,\n",
    "    \"Seed_sequence_common\": str,\n",
    "    \"WELL_QUALITY_DESCRIPTION\": str,\n",
    "    \"siRNA_error\": str,\n",
    "    \"siRNA_number\": str,\n",
    "    \"Precursor_Name\": str\n",
    "}\n",
    "\n",
    "# Dask DataFrames exhibit a peculiarity regarding the index labels: By\n",
    "# default, the index labels are integers, just as with Pandas\n",
    "# DataFrames. However, unlike Pandas DataFrames, the index labels do not\n",
    "# monotonically increase from 0, but restart at 0 for each partition,\n",
    "# thereby resulting in duplicated index labels (Dask subdivides a Dask\n",
    "# DataFram into multiple so-called partitions as the whole idea behind\n",
    "# Dask is to handle large data sets in a memory-efficient way, https://\n",
    "# docs.dask.org/en/stable/generated/dask.dataframe.DataFrame.reset_\n",
    "# index.html)\n",
    "# Hence, performing operations with Dask DataFrames might potentially\n",
    "# raise the `ValueError: cannot reindex on an axis with duplicate\n",
    "# labels` error\n",
    "# In this case, loading the entire data set into a Pandas DataFrame is\n",
    "# feasible, which is why this is preferred to loading it into a Dask\n",
    "# DataFrame (strangely enough, this has not been possible in the very\n",
    "# beginning, which is why Dask was used in the first place)\n",
    "main_csv_df = pd.read_csv(\n",
    "    \"VacciniaReport_20170223-0958_ZScored_conc_and_NaN_adjusted.csv\",\n",
    "    sep=\"\\t\",\n",
    "    dtype=dtype_dict\n",
    ")\n",
    "\n",
    "# Bear in mind that due to operator precedence, i.e. \"|\" (logical OR)\n",
    "# having precedence over equality checks, the equality checks have to be\n",
    "# surrounded by parentheses\n",
    "single_pooled_siRNA_and_esiRNA_df = main_csv_df.loc[\n",
    "    (main_csv_df[\"WellType\"] == \"SIRNA\")\n",
    "    |\n",
    "    (main_csv_df[\"WellType\"] == \"POOLED_SIRNA\")\n",
    "    |\n",
    "    (main_csv_df[\"WellType\"] == \"ESIRNA\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "single_pooled_siRNA_and_esiRNA_IDs = single_pooled_siRNA_and_esiRNA_df[\n",
    "    \"ID_manufacturer\"\n",
    "]\n",
    "\n",
    "assert \"Not available\" not in single_pooled_siRNA_and_esiRNA_IDs, (\n",
    "    \"Contrary to our expectations, the entry \\\"Not available\\\" does \"\n",
    "    \"occur in the feature \\\"ID_manufacturer\\\"!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 21\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;66;03m# Query NCBI's gene database with the Gene ID currently dealt\u001b[39;00m\n\u001b[1;32m     17\u001b[0m \u001b[38;5;66;03m# with\u001b[39;00m\n\u001b[1;32m     18\u001b[0m \u001b[38;5;66;03m# Code execution is suspended for three seconds in order to\u001b[39;00m\n\u001b[1;32m     19\u001b[0m \u001b[38;5;66;03m# avoid server-side errors\u001b[39;00m\n\u001b[1;32m     20\u001b[0m time\u001b[38;5;241m.\u001b[39msleep(\u001b[38;5;241m3\u001b[39m)\n\u001b[0;32m---> 21\u001b[0m NCBI_entry \u001b[38;5;241m=\u001b[39m \u001b[43mentrez\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfetch_single_file\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     22\u001b[0m \u001b[43m    \u001b[49m\u001b[43muids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[43mNCBI_Gene_ID\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     23\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfile_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     24\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdb_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     25\u001b[0m \u001b[43m    \u001b[49m\u001b[43mret_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     26\u001b[0m \u001b[43m    \u001b[49m\u001b[43mret_mode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtext\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\n\u001b[1;32m     27\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     29\u001b[0m \u001b[38;5;66;03m# As the file_name is specified to be None, Biotite's\u001b[39;00m\n\u001b[1;32m     30\u001b[0m \u001b[38;5;66;03m# fetch_single_file() function returns a StringIO object\u001b[39;00m\n\u001b[1;32m     31\u001b[0m \u001b[38;5;66;03m# It's content can be accessed via the getvalue() method\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[38;5;66;03m# that repeatedly using the read() method returns an empty\u001b[39;00m\n\u001b[1;32m     35\u001b[0m \u001b[38;5;66;03m# string\u001b[39;00m\n\u001b[1;32m     36\u001b[0m NCBI_entry_str \u001b[38;5;241m=\u001b[39m NCBI_entry\u001b[38;5;241m.\u001b[39mgetvalue()\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/biotite/database/entrez/download.py:213\u001b[0m, in \u001b[0;36mfetch_single_file\u001b[0;34m(uids, file_name, db_name, ret_type, ret_mode, overwrite)\u001b[0m\n\u001b[1;32m    211\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m api_key \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    212\u001b[0m     param_dict[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mapi_key\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m api_key\n\u001b[0;32m--> 213\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_fetch_url\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparam_dict\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    214\u001b[0m content \u001b[38;5;241m=\u001b[39m r\u001b[38;5;241m.\u001b[39mtext\n\u001b[1;32m    215\u001b[0m check_for_errors(content)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/requests/api.py:73\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     63\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/requests/sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    584\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[1;32m    586\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    587\u001b[0m }\n\u001b[1;32m    588\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 589\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/requests/sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    700\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    705\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    706\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/requests/adapters.py:667\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    664\u001b[0m     timeout \u001b[38;5;241m=\u001b[39m TimeoutSauce(connect\u001b[38;5;241m=\u001b[39mtimeout, read\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[1;32m    666\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 667\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    668\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    669\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    670\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    671\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    672\u001b[0m \u001b[43m        \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    673\u001b[0m \u001b[43m        \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    674\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    675\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    676\u001b[0m \u001b[43m        \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    677\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    678\u001b[0m \u001b[43m        \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    679\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    681\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ProtocolError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    682\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(err, request\u001b[38;5;241m=\u001b[39mrequest)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/urllib3/connectionpool.py:793\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, preload_content, decode_content, **response_kw)\u001b[0m\n\u001b[1;32m    790\u001b[0m response_conn \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m release_conn \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    792\u001b[0m \u001b[38;5;66;03m# Make the request on the HTTPConnection object\u001b[39;00m\n\u001b[0;32m--> 793\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    794\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    795\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    796\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    797\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    798\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    799\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    800\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    801\u001b[0m \u001b[43m    \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    802\u001b[0m \u001b[43m    \u001b[49m\u001b[43mresponse_conn\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mresponse_conn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    803\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpreload_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    804\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecode_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    805\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mresponse_kw\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    806\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    808\u001b[0m \u001b[38;5;66;03m# Everything went great!\u001b[39;00m\n\u001b[1;32m    809\u001b[0m clean_exit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/urllib3/connectionpool.py:537\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, body, headers, retries, timeout, chunked, response_conn, preload_content, decode_content, enforce_content_length)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[38;5;66;03m# Receive the response from the server\u001b[39;00m\n\u001b[1;32m    536\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 537\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresponse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    538\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (BaseSSLError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    539\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raise_timeout(err\u001b[38;5;241m=\u001b[39me, url\u001b[38;5;241m=\u001b[39murl, timeout_value\u001b[38;5;241m=\u001b[39mread_timeout)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/urllib3/connection.py:466\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    463\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mresponse\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m HTTPResponse\n\u001b[1;32m    465\u001b[0m \u001b[38;5;66;03m# Get the response from http.client.HTTPConnection\u001b[39;00m\n\u001b[0;32m--> 466\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresponse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    468\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    469\u001b[0m     assert_header_parsing(httplib_response\u001b[38;5;241m.\u001b[39mmsg)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/http/client.py:1374\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1372\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1373\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1374\u001b[0m         \u001b[43mresponse\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbegin\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1375\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m:\n\u001b[1;32m   1376\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/http/client.py:318\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    316\u001b[0m \u001b[38;5;66;03m# read until we get a non-100 response\u001b[39;00m\n\u001b[1;32m    317\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 318\u001b[0m     version, status, reason \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_read_status\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    319\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m status \u001b[38;5;241m!=\u001b[39m CONTINUE:\n\u001b[1;32m    320\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/http/client.py:279\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_read_status\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 279\u001b[0m     line \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreadline\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_MAXLINE\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124miso-8859-1\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    280\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(line) \u001b[38;5;241m>\u001b[39m _MAXLINE:\n\u001b[1;32m    281\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m LineTooLong(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstatus line\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/socket.py:705\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    703\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m    704\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 705\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv_into\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    706\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[1;32m    707\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/ssl.py:1274\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1270\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m flags \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   1271\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1272\u001b[0m           \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[1;32m   1273\u001b[0m           \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m)\n\u001b[0;32m-> 1274\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnbytes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1275\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1276\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mrecv_into(buffer, nbytes, flags)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/ssl.py:1130\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1128\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1129\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m buffer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1130\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1131\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1132\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "ID_change_str = \"This record was replaced with GeneID:\"\n",
    "\n",
    "# Iterate over the DataFrame rows using the iterrows() method\n",
    "for i, row in main_csv_df.iterrows():\n",
    "    # Incorporating kind of a numeric progress bar\n",
    "    if i % 10000 == 0:\n",
    "        print(i)\n",
    "    well_type = row[\"WellType\"]\n",
    "    if (\n",
    "        well_type == \"SIRNA\"\n",
    "        or\n",
    "        well_type == \"POOLED_SIRNA\"\n",
    "        or\n",
    "        well_type == \"ESIRNA\"\n",
    "    ):\n",
    "        gene_name = row[\"Name\"]\n",
    "        NCBI_Gene_ID = row[\"ID_manufacturer\"]\n",
    "        \n",
    "        # Query NCBI's gene database with the Gene ID currently dealt\n",
    "        # with\n",
    "        # Code execution is suspended for two seconds in order to avoid\n",
    "        # server-side errors\n",
    "        # The VACV data set comprises 132,066 measurements involving\n",
    "        # single siRNAs, pooled siRNAs and esiRNAs\n",
    "        # For each of those measurements, the NCBI database is queried\n",
    "        # at least once, entailing the suspension of code execution for\n",
    "        # two seconds; in total, this amounts to a \"waiting time\" of\n",
    "        # 264,132 seconds, which corresponds to slightly more than three\n",
    "        # days; this period of time is acceptable\n",
    "        time.sleep(2)\n",
    "        NCBI_entry = entrez.fetch_single_file(\n",
    "            uids=[NCBI_Gene_ID],\n",
    "            file_name=None,\n",
    "            db_name=\"gene\",\n",
    "            ret_type=\"\",\n",
    "            ret_mode=\"text\"\n",
    "        )\n",
    "\n",
    "        # As the file_name is specified to be None, Biotite's\n",
    "        # fetch_single_file() function returns a StringIO object\n",
    "        # It's content can be accessed via the getvalue() method\n",
    "        # Note that the getvalue() method is preferred to the read()\n",
    "        # method as the latter moves the cursor to the last index so\n",
    "        # that repeatedly using the read() method returns an empty\n",
    "        # string\n",
    "        NCBI_entry_str = NCBI_entry.getvalue()\n",
    "\n",
    "        # Different approaches are necessary depending on whether merely\n",
    "        # the official gene symbol was altered, whereas the NCBI Gene ID\n",
    "        # remained unchanged, or the record has been replaced altogether\n",
    "        # with an entirely new ID\n",
    "        if ID_change_str in NCBI_entry_str:\n",
    "            # The respective record has been replaced altogether with a\n",
    "            # new ID\n",
    "            # Hence, the new ID is retrieved and used to query NCBI's\n",
    "            # gene database\n",
    "            NCBI_entry_str_list = NCBI_entry_str.split(\"\\n\")\n",
    "            # For some strange reason, the string retrieved from the\n",
    "            # NCBI entry contains blank lines; they are removed\n",
    "            while \"\" in NCBI_entry_str_list:\n",
    "                NCBI_entry_str_list.remove(\"\")\n",
    "            \n",
    "            # The new ID is comprised in the penultimate list element\n",
    "            # and conveniently enough separated from the preceding\n",
    "            # string by a space character\n",
    "            new_gene_ID = NCBI_entry_str_list[-1].split()[-1]\n",
    "            main_csv_df.at[i, \"ID\"] = new_gene_ID\n",
    "            main_csv_df.at[i, \"ID_manufacturer\"] = new_gene_ID\n",
    "\n",
    "            # Again, in a bid to prevent the occurrence of server-side\n",
    "            # errors, code execution is suspended for two seconds\n",
    "            time.sleep(2)\n",
    "            NCBI_entry = entrez.fetch_single_file(\n",
    "                uids=[new_gene_ID],\n",
    "                file_name=None,\n",
    "                db_name=\"gene\",\n",
    "                ret_type=\"\",\n",
    "                ret_mode=\"text\"\n",
    "            )\n",
    "            NCBI_entry_str = NCBI_entry.getvalue()\n",
    "            NCBI_entry_str_list = NCBI_entry_str.split(\"\\n\")\n",
    "            while \"\" in NCBI_entry_str_list:\n",
    "                NCBI_entry_str_list.remove(\"\")\n",
    "            \n",
    "            # The official gene symbol is comprised in the first list\n",
    "            # element, but is preceded by the string \"1. \",\n",
    "            # which encompasses three characters\n",
    "            official_gene_symbol = NCBI_entry_str_list[0][3:]\n",
    "\n",
    "            main_csv_df.at[i, \"Name\"] = official_gene_symbol\n",
    "        else:\n",
    "            # Remove blank lines from the string retrieved from the NCBI\n",
    "            # entry\n",
    "            NCBI_entry_str_list = NCBI_entry_str.split(\"\\n\")\n",
    "            while \"\" in NCBI_entry_str_list:\n",
    "                NCBI_entry_str_list.remove(\"\")\n",
    "            \n",
    "            # Following the removal of empty strings, the official gene\n",
    "            # symbol is represented by the first list element, but it is\n",
    "            # preceded by the string \"1. \", which encompasses three\n",
    "            # characters\n",
    "            # Hence, prior to comparing the gene names provided by the\n",
    "            # VACV screen data set to the official gene symbols, the\n",
    "            # first list element has to be sliced accordingly\n",
    "            official_gene_symbol = NCBI_entry_str_list[0][3:]\n",
    "            if gene_name != official_gene_symbol:\n",
    "                main_csv_df.at[i, \"Name\"] = official_gene_symbol\n",
    "            \n",
    "            # Irrespective of whether the gene name provided by the VACV\n",
    "            # data set and the official gene symbol match or not, the\n",
    "            # corresponding cell of the \"ID\" feature is populated with\n",
    "            # the NCBI Gene ID harboured by the \"ID_manufacturer\"\n",
    "            # feature (remember that cells of the \"ID\" feature are not\n",
    "            # continuously populated)\n",
    "            main_csv_df.at[i, \"ID\"] = NCBI_Gene_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# As a last step, save the new Pandas DataFrame to a new CSV file\n",
    "main_csv_df.to_csv(\n",
    "    \"Vaccinia_Report_NCBI_Gene_IDs_and_official_gene_symbols_updated.csv\",\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Human-virus PPI', 'Experimental System', 'PubMed ID', 'Virus acronym'], dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# A UniProt reference proteome has been found, which encompasses 218\n",
    "# proteins\n",
    "# It is checked whether these 218 proteins comprise all interacting VACV\n",
    "# proteins recorded by HVIDB or whether HVIDB contains VACV proteins not\n",
    "# comprised in the reference proteome\n",
    "# First, perform a numerical check\n",
    "HVIDB_VACV_interactions_df = pd.read_csv(\n",
    "    \"all_HVIDB_VACV_interactions.csv\"\n",
    ")\n",
    "print(HVIDB_VACV_interactions_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The VACV UniProt IDs are comprised in the first column bearing the\n",
    "# name \"Human-virus PPI\"; to be more precise, the first column\n",
    "# represents an interaction pair between a human and a VACV protein as a\n",
    "# combination of the respective UniProt IDs, which are separated by a\n",
    "# hyphen\n",
    "# The first UniProt ID represents a human protein, whereas the second\n",
    "# UniProt ID represents a VACV protein\n",
    "interaction_pair_IDs = HVIDB_VACV_interactions_df[\n",
    "    \"Human-virus PPI\"\n",
    "].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "VACV_interaction_IDs = [\n",
    "    pair.split(\"-\")[1] for pair in interaction_pair_IDs\n",
    "]\n",
    "\n",
    "assert len(interaction_pair_IDs) == len(VACV_interaction_IDs), (\n",
    "    \"Something went wrong!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bear in mind that one and the same protein can engage in more than one\n",
    "# PPI, which also applies to VACV\n",
    "# Hence, duplicate UniProt IDs are discarded\n",
    "VACV_unique_interaction_IDs = np.unique(VACV_interaction_IDs).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61\n"
     ]
    }
   ],
   "source": [
    "print(len(VACV_unique_interaction_IDs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 64858 DCLRE1B\n",
    "# Incorporate time.sleep() in order to avoid the error?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A0 in Inkscape\n",
    "# Prospective Poster\n",
    "# Describe Task, Setting, Methods, intermediate results (e.g. XGBoost)\n",
    "# Use HZDR corporate Design/CASUS template for poster (both HZDR and CASUS)\n",
    "# Regarding compounds and siRNA: they are different, but one could be used\n",
    "# to validate the other\n",
    "# Exactly the same effect is not expected"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinformatics",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
