{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "import math\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from biotite.database import entrez\n",
    "from biotite.database import uniprot\n",
    "import biotite.sequence.io.fasta as fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unfortunately, to my knowledge, downloading all VACV interactions from\n",
    "# the HVIDB database at once is not possible\n",
    "# Hence, they have been downloaded in packages of up to 100 interactions\n",
    "# It is investigated whether the 5 packages are all different from each\n",
    "# other so as to rule out errors during the manual download, such as\n",
    "# accidentally downloading the same package twice\n",
    "# As a first step, the individual packages are loaded into the Jupyter\n",
    "# notebook\n",
    "HVIDB_package_1 = pd.read_csv(\"HVIDB_VACV_interactions_1.csv\")\n",
    "HVIDB_package_2 = pd.read_csv(\"HVIDB_VACV_interactions_2.csv\")\n",
    "HVIDB_package_3 = pd.read_csv(\"HVIDB_VACV_interactions_3.csv\")\n",
    "HVIDB_package_4 = pd.read_csv(\"HVIDB_VACV_interactions_4.csv\")\n",
    "HVIDB_package_5 = pd.read_csv(\"HVIDB_VACV_interactions_5.csv\")\n",
    "\n",
    "# According to the HVIDB database, the total amount of interactions is\n",
    "# 456\n",
    "# It is checked whether this also applies to the five downloaded\n",
    "# packages\n",
    "packages_list = [\n",
    "    HVIDB_package_1,\n",
    "    HVIDB_package_2,\n",
    "    HVIDB_package_3,\n",
    "    HVIDB_package_4,\n",
    "    HVIDB_package_5\n",
    "]\n",
    "\n",
    "assert sum(map(len, packages_list)) == 456, (\n",
    "    \"An error has been made during the manual download of the \"\n",
    "    \"interaction packages!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, verify that all downloaded interaction packages are distinct from\n",
    "# one another\n",
    "# Unfortunately, checking equality for more than two DataFrames at once\n",
    "# is not possible to the best of my knowledge\n",
    "# Hence, it is resorted to for-loops\n",
    "package_equality = False\n",
    "\n",
    "for i, package in enumerate(packages_list):\n",
    "    # For obvious reasons, equality checks are only performed between\n",
    "    # different interaction packages\n",
    "    # Moreover, redundant equality checks are avoided\n",
    "    for j in range(i + 1, 5):\n",
    "        package_equality = package_equality or package.equals(packages_list[j])\n",
    "\n",
    "assert package_equality == False, (\n",
    "    \"Identical interaction packages have erroneously been downloaded!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, for the sake of convenience, concatenate the five individual\n",
    "# interaction packages to generate one coherent CSV file\n",
    "HVIDB_VACV_interactions_df = pd.concat(packages_list, ignore_index=True)\n",
    "\n",
    "assert (\n",
    "    len(HVIDB_VACV_interactions_df) == 456\n",
    "    and\n",
    "    len(HVIDB_VACV_interactions_df.columns) == 4\n",
    "), \"A mistake was done during DataFrame concatenation!\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the coherent DataFrame as a CSV file\n",
    "HVIDB_VACV_interactions_df.to_csv(\n",
    "    \"all_HVIDB_VACV_interactions.csv\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39\n",
      "['11129635' '11238845' '11878931' '12566418' '12960323' '14963148'\n",
      " '15215253' '15518812' '15767367' '15998638' '16003387' '1610791'\n",
      " '16254338' '16439990' '16840345' '17074758' '17123957' '17166913'\n",
      " '17485524' '17681535' '18005740' '18208323' '18266467' '18353424'\n",
      " '18551131' '18604270' '18636090' '18971339' '19637933' '19913487'\n",
      " '21915095' '21931555' '22810585' '24550280' '7609027' '9199350' '9431994'\n",
      " '9792841' '9824158']\n"
     ]
    }
   ],
   "source": [
    "# Determine the unique values for the column \"PubMed ID\", i.e. determine\n",
    "# the publications interaction information was retrieved from\n",
    "pre_unique_pubs = np.unique(HVIDB_VACV_interactions_df[\"PubMed ID\"])\n",
    "# Some cells contain multiple PubMed IDs separated by commata, which\n",
    "# requires special handling\n",
    "pre_unique_single_pubs = []\n",
    "for pub_id in pre_unique_pubs:\n",
    "    if \",\" in pub_id:\n",
    "        pre_unique_single_pubs += pub_id.split(\",\")\n",
    "    else:\n",
    "        pre_unique_single_pubs.append(pub_id)\n",
    "unique_pubs = np.unique(pre_unique_single_pubs)\n",
    "print(len(unique_pubs))\n",
    "print(unique_pubs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amount of human proteins targeted in the VACV screen by single/pooled siRNAs as well as esiRNAs: 18,395\n"
     ]
    }
   ],
   "source": [
    "# Determine the amount of genes targeted by siRNAs/esiRNAs in the VACV\n",
    "# screen\n",
    "# Low-molecular weight organic compounds are not taken into account as\n",
    "# despite having the same final outcome as siRNAs, i.e. the inhibition\n",
    "# of the action of proteins, low-molecular weight compounds and siRNAs\n",
    "# differ from one another in their mode of action: The former come into\n",
    "# play after protein translation has taken place, whereas the latter\n",
    "# unfold their effect on the post-transcriptional level by targeting\n",
    "# mRNAs\n",
    "# Hence, exactly the same effect is not expected\n",
    "single_pooled_siRNA_and_esiRNA_STRING_IDs_df = pd.read_csv(\n",
    "    \"single_pooled_siRNA_and_esiRNA_STRING_IDs.csv\",\n",
    "    delimiter=\"\\t\"\n",
    ")\n",
    "\n",
    "n_targets = len(single_pooled_siRNA_and_esiRNA_STRING_IDs_df)\n",
    "\n",
    "print(\n",
    "    \"Amount of human proteins targeted in the VACV screen by \"\n",
    "    f\"single/pooled siRNAs as well as esiRNAs: {n_targets:,}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, determine the amount of human genes interrogated in the VACV\n",
    "# screen that engage in PPIs recorded by HVIDB\n",
    "# To this end, the protein target names have to be converted in\n",
    "# UniProtIDs as they are employed by HVIDB\n",
    "# This is achieved via the ID mapping GUI on\n",
    "# https://www.uniprot.org/id-mapping\n",
    "VACV_screen_human_targets_series = (\n",
    "    single_pooled_siRNA_and_esiRNA_STRING_IDs_df[\"queryItem\"]\n",
    ")\n",
    "\n",
    "VACV_screen_human_targets_series.to_csv(\n",
    "    \"VACV_screen_human_targets_single_pooled_siRNA_and_esiRNA_gene_names.txt\",\n",
    "    header=False,\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using the parameters \"from: UniProt/Gene Name\",\n",
    "# \"to: UniProtKB-Swiss-Prot\" and \"taxId: 9606\", 18,185 of the 18,396\n",
    "# human protein IDs were mapped to 18,645 results, whereas 213 ID were\n",
    "# not mapped\n",
    "# This is subjected to close scrutiny\n",
    "#\n",
    "# For the ALG1L gene, the official gene symbol has apparently been\n",
    "# changed after the VACV screen has been conducted; its new official\n",
    "# gene symbol is ALG1L1P while its NCBI Gene ID remained unchanged\n",
    "# (200810)\n",
    "#\n",
    "# Regarding the gene AKAP2, the NCBI Gene ID has been changed from 11217\n",
    "# to 445815; note that AKAP2 is only an alias and that the offical gene\n",
    "# symbol is PALM2AKAP2\n",
    "#\n",
    "# With respect to the gene PPAN-P2RY11, neither the official gene symbol\n",
    "# nor the NCBI Gene ID have been altered after the VACV screen, but the\n",
    "# respective UniProtKB entry belongs to the UniProtKB/TrEMBL section,\n",
    "# i.e. has not been manually reviewed. Consequently, using the setting\n",
    "# \"to: UniProtKB-Swiss-Prot\" does not yield any UniProtID for this gene\n",
    "# It becomes apparent that the most convenient way of obtaining\n",
    "# UniProtIDs in one single step is to use the settings\n",
    "# \"From database: Genomic annotation databases/GeneID\" (Gene ID is the\n",
    "# same as NCBI Gene ID) and \"To database: UniProt/UniProtKB\" (choosing\n",
    "# UniProtKB includes both UniProtKB sections, i.e. UniProtKB/Swiss-Prot\n",
    "# and UniProtKB/TrEMBL)\n",
    "#\n",
    "# With regard to the gene CDY, the specified NCBI Gene ID is still valid\n",
    "# (203611), but its official gene symbol is CDY2B\n",
    "#\n",
    "# As to the gene FLJ40296, the specified NCBI Gene ID is still valid\n",
    "# (122183), but its official gene symbol is PRR20A\n",
    "#\n",
    "# Regarding the gene MGC39606, the specified NCBI Gene ID is still valid\n",
    "# (399668), but its official gene symbol is SMIM10L2A\n",
    "#\n",
    "# etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Instead of doing this investigation manually, it is accomplished in an\n",
    "# automated fashion by means of the Biotite bioinformatics library\n",
    "# Biotite provides functionalities allowing to access various NCBI\n",
    "# databases in a programmatic manner, i.e. from within Python scripts\n",
    "\n",
    "# Two columns of interest for the following endeavour are \"ID\" as well\n",
    "# as \"ID_manufacturer\"\n",
    "# Upon closer scrutiny, it became apparent that contrary to their names,\n",
    "# both features basically harbour the same information, namely the NCBI\n",
    "# Gene ID of the targeted gene\n",
    "# However, they differ in that while the feature \"ID\" is populated\n",
    "# depending on whether the respective experiment was successful or not,\n",
    "# \"ID_manufacturer\" is continuously populated, irrespective of the\n",
    "# experiment's outcome\n",
    "# Hence, the entries of the \"ID_manufacturer\" feature are employed in\n",
    "# order to query NCBI's gene database\n",
    "# As a first step, it is checked whether all entries of the\n",
    "# \"ID_manufacturer\" feature are indeed populated with values or whether\n",
    "# there are isolated occurrences of \"Not available\"\n",
    "# Note that the features \"ID\" and \"ID_manufacturer\" only harbour NCBI\n",
    "# Gene IDs in the case of siRNA, pooled siRNA and esiRNA\n",
    "\n",
    "# Load the screen data\n",
    "# Bear in mind that for certain columns, the data type has to be\n",
    "# manually specified\n",
    "dtype_dict = {\n",
    "    \"Ensembl_ID_OnTarget_Ensembl_GRCh38_release_87\": str,\n",
    "    \"Ensembl_ID_OnTarget_NCBI_HeLa_phs000643_v3_p1_c1_HMB\": str,\n",
    "    \"Gene_Description\": str,\n",
    "    \"ID\": str,\n",
    "    \"ID_OnTarget_Ensembl_GRCh38_release_87\": str,\n",
    "    \"ID_OnTarget_Merge\": str,\n",
    "    \"ID_OnTarget_NCBI_HeLa_phs000643_v3_p1_c1_HMB\": str,\n",
    "    \"ID_OnTarget_RefSeq_20170215\": str,\n",
    "    \"ID_manufacturer\": str,\n",
    "    \"Name_alternatives\": str,\n",
    "    \"PLATE_QUALITY_DESCRIPTION\": str,\n",
    "    \"RefSeq_ID_OnTarget_RefSeq_20170215\": str,\n",
    "    \"Seed_sequence_common\": str,\n",
    "    \"WELL_QUALITY_DESCRIPTION\": str,\n",
    "    \"siRNA_error\": str,\n",
    "    \"siRNA_number\": str,\n",
    "    \"Precursor_Name\": str\n",
    "}\n",
    "\n",
    "# Dask DataFrames exhibit a peculiarity regarding the index labels: By\n",
    "# default, the index labels are integers, just as with Pandas\n",
    "# DataFrames. However, unlike Pandas DataFrames, the index labels do not\n",
    "# monotonically increase from 0, but restart at 0 for each partition,\n",
    "# thereby resulting in duplicated index labels (Dask subdivides a Dask\n",
    "# DataFram into multiple so-called partitions as the whole idea behind\n",
    "# Dask is to handle large data sets in a memory-efficient way, https://\n",
    "# docs.dask.org/en/stable/generated/dask.dataframe.DataFrame.reset_\n",
    "# index.html)\n",
    "# Hence, performing operations with Dask DataFrames might potentially\n",
    "# raise the `ValueError: cannot reindex on an axis with duplicate\n",
    "# labels` error\n",
    "# In this case, loading the entire data set into a Pandas DataFrame is\n",
    "# feasible, which is why this is preferred to loading it into a Dask\n",
    "# DataFrame (strangely enough, this has not been possible in the very\n",
    "# beginning, which is why Dask was used in the first place)\n",
    "main_csv_df = pd.read_csv(\n",
    "    \"VacciniaReport_20170223-0958_ZScored_conc_and_NaN_adjusted.csv\",\n",
    "    sep=\"\\t\",\n",
    "    dtype=dtype_dict\n",
    ")\n",
    "\n",
    "# Bear in mind that due to operator precedence, i.e. \"|\" (logical OR)\n",
    "# having precedence over equality checks, the equality checks have to be\n",
    "# surrounded by parentheses\n",
    "single_pooled_siRNA_and_esiRNA_df = main_csv_df.loc[\n",
    "    (main_csv_df[\"WellType\"] == \"SIRNA\")\n",
    "    |\n",
    "    (main_csv_df[\"WellType\"] == \"POOLED_SIRNA\")\n",
    "    |\n",
    "    (main_csv_df[\"WellType\"] == \"ESIRNA\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "Contrary to our expectations, the entry \"Not available\" does occur in the feature \"ID_manufacturer\"!",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 10\u001b[0m\n\u001b[1;32m      1\u001b[0m single_pooled_siRNA_and_esiRNA_IDs \u001b[38;5;241m=\u001b[39m single_pooled_siRNA_and_esiRNA_df[\n\u001b[1;32m      2\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mID_manufacturer\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      3\u001b[0m ]\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# Bear in mind that the in operator cannot be used in the context of\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;66;03m# Pandas Serieses as it refers to the index instead of the data itself\u001b[39;00m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m# Thus, the `in` operator is employed in conjunction with the `values`\u001b[39;00m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;66;03m# attribute of Pandas Serieses, which returns the Series as a NumPy\u001b[39;00m\n\u001b[1;32m      9\u001b[0m \u001b[38;5;66;03m# array\u001b[39;00m\n\u001b[0;32m---> 10\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNot available\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m single_pooled_siRNA_and_esiRNA_IDs\u001b[38;5;241m.\u001b[39mvalues, (\n\u001b[1;32m     11\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mContrary to our expectations, the entry \u001b[39m\u001b[38;5;130;01m\\\"\u001b[39;00m\u001b[38;5;124mNot available\u001b[39m\u001b[38;5;130;01m\\\"\u001b[39;00m\u001b[38;5;124m does \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     12\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124moccur in the feature \u001b[39m\u001b[38;5;130;01m\\\"\u001b[39;00m\u001b[38;5;124mID_manufacturer\u001b[39m\u001b[38;5;130;01m\\\"\u001b[39;00m\u001b[38;5;124m!\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     13\u001b[0m )\n",
      "\u001b[0;31mAssertionError\u001b[0m: Contrary to our expectations, the entry \"Not available\" does occur in the feature \"ID_manufacturer\"!"
     ]
    }
   ],
   "source": [
    "single_pooled_siRNA_and_esiRNA_IDs = single_pooled_siRNA_and_esiRNA_df[\n",
    "    \"ID_manufacturer\"\n",
    "]\n",
    "\n",
    "# Bear in mind that the in operator cannot be used in the context of\n",
    "# Pandas Serieses as it refers to the index instead of the data itself\n",
    "# Thus, the `in` operator is employed in conjunction with the `values`\n",
    "# attribute of Pandas Serieses, which returns the Series as a NumPy\n",
    "# array\n",
    "assert \"Not available\" not in single_pooled_siRNA_and_esiRNA_IDs.values, (\n",
    "    \"Contrary to our expectations, the entry \\\"Not available\\\" does \"\n",
    "    \"occur in the feature \\\"ID_manufacturer\\\"!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 21\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;66;03m# Query NCBI's gene database with the Gene ID currently dealt\u001b[39;00m\n\u001b[1;32m     17\u001b[0m \u001b[38;5;66;03m# with\u001b[39;00m\n\u001b[1;32m     18\u001b[0m \u001b[38;5;66;03m# Code execution is suspended for three seconds in order to\u001b[39;00m\n\u001b[1;32m     19\u001b[0m \u001b[38;5;66;03m# avoid server-side errors\u001b[39;00m\n\u001b[1;32m     20\u001b[0m time\u001b[38;5;241m.\u001b[39msleep(\u001b[38;5;241m3\u001b[39m)\n\u001b[0;32m---> 21\u001b[0m NCBI_entry \u001b[38;5;241m=\u001b[39m \u001b[43mentrez\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfetch_single_file\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     22\u001b[0m \u001b[43m    \u001b[49m\u001b[43muids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[43mNCBI_Gene_ID\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     23\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfile_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     24\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdb_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     25\u001b[0m \u001b[43m    \u001b[49m\u001b[43mret_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     26\u001b[0m \u001b[43m    \u001b[49m\u001b[43mret_mode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtext\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\n\u001b[1;32m     27\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     29\u001b[0m \u001b[38;5;66;03m# As the file_name is specified to be None, Biotite's\u001b[39;00m\n\u001b[1;32m     30\u001b[0m \u001b[38;5;66;03m# fetch_single_file() function returns a StringIO object\u001b[39;00m\n\u001b[1;32m     31\u001b[0m \u001b[38;5;66;03m# It's content can be accessed via the getvalue() method\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[38;5;66;03m# that repeatedly using the read() method returns an empty\u001b[39;00m\n\u001b[1;32m     35\u001b[0m \u001b[38;5;66;03m# string\u001b[39;00m\n\u001b[1;32m     36\u001b[0m NCBI_entry_str \u001b[38;5;241m=\u001b[39m NCBI_entry\u001b[38;5;241m.\u001b[39mgetvalue()\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/biotite/database/entrez/download.py:213\u001b[0m, in \u001b[0;36mfetch_single_file\u001b[0;34m(uids, file_name, db_name, ret_type, ret_mode, overwrite)\u001b[0m\n\u001b[1;32m    211\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m api_key \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    212\u001b[0m     param_dict[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mapi_key\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m api_key\n\u001b[0;32m--> 213\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_fetch_url\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparam_dict\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    214\u001b[0m content \u001b[38;5;241m=\u001b[39m r\u001b[38;5;241m.\u001b[39mtext\n\u001b[1;32m    215\u001b[0m check_for_errors(content)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/requests/api.py:73\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     63\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/requests/sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    584\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[1;32m    586\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    587\u001b[0m }\n\u001b[1;32m    588\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 589\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/requests/sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    700\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    705\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    706\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/requests/adapters.py:667\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    664\u001b[0m     timeout \u001b[38;5;241m=\u001b[39m TimeoutSauce(connect\u001b[38;5;241m=\u001b[39mtimeout, read\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[1;32m    666\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 667\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    668\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    669\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    670\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    671\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    672\u001b[0m \u001b[43m        \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    673\u001b[0m \u001b[43m        \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    674\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    675\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    676\u001b[0m \u001b[43m        \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    677\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    678\u001b[0m \u001b[43m        \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    679\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    681\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ProtocolError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    682\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(err, request\u001b[38;5;241m=\u001b[39mrequest)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/urllib3/connectionpool.py:793\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, preload_content, decode_content, **response_kw)\u001b[0m\n\u001b[1;32m    790\u001b[0m response_conn \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m release_conn \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    792\u001b[0m \u001b[38;5;66;03m# Make the request on the HTTPConnection object\u001b[39;00m\n\u001b[0;32m--> 793\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    794\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    795\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    796\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    797\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    798\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    799\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    800\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    801\u001b[0m \u001b[43m    \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    802\u001b[0m \u001b[43m    \u001b[49m\u001b[43mresponse_conn\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mresponse_conn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    803\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpreload_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    804\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecode_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    805\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mresponse_kw\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    806\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    808\u001b[0m \u001b[38;5;66;03m# Everything went great!\u001b[39;00m\n\u001b[1;32m    809\u001b[0m clean_exit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/urllib3/connectionpool.py:537\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, body, headers, retries, timeout, chunked, response_conn, preload_content, decode_content, enforce_content_length)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[38;5;66;03m# Receive the response from the server\u001b[39;00m\n\u001b[1;32m    536\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 537\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresponse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    538\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (BaseSSLError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    539\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raise_timeout(err\u001b[38;5;241m=\u001b[39me, url\u001b[38;5;241m=\u001b[39murl, timeout_value\u001b[38;5;241m=\u001b[39mread_timeout)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/urllib3/connection.py:466\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    463\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mresponse\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m HTTPResponse\n\u001b[1;32m    465\u001b[0m \u001b[38;5;66;03m# Get the response from http.client.HTTPConnection\u001b[39;00m\n\u001b[0;32m--> 466\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresponse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    468\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    469\u001b[0m     assert_header_parsing(httplib_response\u001b[38;5;241m.\u001b[39mmsg)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/http/client.py:1374\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1372\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1373\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1374\u001b[0m         \u001b[43mresponse\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbegin\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1375\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m:\n\u001b[1;32m   1376\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/http/client.py:318\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    316\u001b[0m \u001b[38;5;66;03m# read until we get a non-100 response\u001b[39;00m\n\u001b[1;32m    317\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 318\u001b[0m     version, status, reason \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_read_status\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    319\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m status \u001b[38;5;241m!=\u001b[39m CONTINUE:\n\u001b[1;32m    320\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/http/client.py:279\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_read_status\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 279\u001b[0m     line \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreadline\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_MAXLINE\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124miso-8859-1\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    280\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(line) \u001b[38;5;241m>\u001b[39m _MAXLINE:\n\u001b[1;32m    281\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m LineTooLong(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstatus line\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/socket.py:705\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    703\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m    704\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 705\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv_into\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    706\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[1;32m    707\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/ssl.py:1274\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1270\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m flags \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   1271\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1272\u001b[0m           \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[1;32m   1273\u001b[0m           \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m)\n\u001b[0;32m-> 1274\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnbytes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1275\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1276\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mrecv_into(buffer, nbytes, flags)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/ssl.py:1130\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1128\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1129\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m buffer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1130\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1131\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1132\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "ID_change_str = \"This record was replaced with GeneID:\"\n",
    "\n",
    "# Iterate over the DataFrame rows using the iterrows() method\n",
    "for i, row in main_csv_df.iterrows():\n",
    "    # Incorporating kind of a numeric progress bar\n",
    "    if i % 10000 == 0:\n",
    "        print(i)\n",
    "    well_type = row[\"WellType\"]\n",
    "    if (\n",
    "        well_type == \"SIRNA\"\n",
    "        or\n",
    "        well_type == \"POOLED_SIRNA\"\n",
    "        or\n",
    "        well_type == \"ESIRNA\"\n",
    "    ):\n",
    "        gene_name = row[\"Name\"]\n",
    "        NCBI_Gene_ID = row[\"ID_manufacturer\"]\n",
    "        \n",
    "        # Query NCBI's gene database with the Gene ID currently dealt\n",
    "        # with\n",
    "        # Code execution is suspended for two seconds in order to avoid\n",
    "        # server-side errors\n",
    "        # The VACV data set comprises 132,066 measurements involving\n",
    "        # single siRNAs, pooled siRNAs and esiRNAs\n",
    "        # For each of those measurements, the NCBI database is queried\n",
    "        # at least once, entailing the suspension of code execution for\n",
    "        # two seconds; in total, this amounts to a \"waiting time\" of\n",
    "        # 264,132 seconds, which corresponds to slightly more than three\n",
    "        # days; this period of time is acceptable\n",
    "        time.sleep(2)\n",
    "        NCBI_entry = entrez.fetch_single_file(\n",
    "            uids=[NCBI_Gene_ID],\n",
    "            file_name=None,\n",
    "            db_name=\"gene\",\n",
    "            ret_type=\"\",\n",
    "            ret_mode=\"text\"\n",
    "        )\n",
    "\n",
    "        # As the file_name is specified to be None, Biotite's\n",
    "        # fetch_single_file() function returns a StringIO object\n",
    "        # It's content can be accessed via the getvalue() method\n",
    "        # Note that the getvalue() method is preferred to the read()\n",
    "        # method as the latter moves the cursor to the last index so\n",
    "        # that repeatedly using the read() method returns an empty\n",
    "        # string\n",
    "        NCBI_entry_str = NCBI_entry.getvalue()\n",
    "\n",
    "        # Different approaches are necessary depending on whether merely\n",
    "        # the official gene symbol was altered, whereas the NCBI Gene ID\n",
    "        # remained unchanged, or the record has been replaced altogether\n",
    "        # with an entirely new ID\n",
    "        if ID_change_str in NCBI_entry_str:\n",
    "            # The respective record has been replaced altogether with a\n",
    "            # new ID\n",
    "            # Hence, the new ID is retrieved and used to query NCBI's\n",
    "            # gene database\n",
    "            NCBI_entry_str_list = NCBI_entry_str.split(\"\\n\")\n",
    "            # For some strange reason, the string retrieved from the\n",
    "            # NCBI entry contains blank lines; they are removed\n",
    "            while \"\" in NCBI_entry_str_list:\n",
    "                NCBI_entry_str_list.remove(\"\")\n",
    "            \n",
    "            # The new ID is comprised in the penultimate list element\n",
    "            # and conveniently enough separated from the preceding\n",
    "            # string by a space character\n",
    "            new_gene_ID = NCBI_entry_str_list[-1].split()[-1]\n",
    "            main_csv_df.at[i, \"ID\"] = new_gene_ID\n",
    "            main_csv_df.at[i, \"ID_manufacturer\"] = new_gene_ID\n",
    "\n",
    "            # Again, in a bid to prevent the occurrence of server-side\n",
    "            # errors, code execution is suspended for two seconds\n",
    "            time.sleep(2)\n",
    "            NCBI_entry = entrez.fetch_single_file(\n",
    "                uids=[new_gene_ID],\n",
    "                file_name=None,\n",
    "                db_name=\"gene\",\n",
    "                ret_type=\"\",\n",
    "                ret_mode=\"text\"\n",
    "            )\n",
    "            NCBI_entry_str = NCBI_entry.getvalue()\n",
    "            NCBI_entry_str_list = NCBI_entry_str.split(\"\\n\")\n",
    "            while \"\" in NCBI_entry_str_list:\n",
    "                NCBI_entry_str_list.remove(\"\")\n",
    "            \n",
    "            # The official gene symbol is comprised in the first list\n",
    "            # element, but is preceded by the string \"1. \",\n",
    "            # which encompasses three characters\n",
    "            official_gene_symbol = NCBI_entry_str_list[0][3:]\n",
    "\n",
    "            main_csv_df.at[i, \"Name\"] = official_gene_symbol\n",
    "        else:\n",
    "            # Remove blank lines from the string retrieved from the NCBI\n",
    "            # entry\n",
    "            NCBI_entry_str_list = NCBI_entry_str.split(\"\\n\")\n",
    "            while \"\" in NCBI_entry_str_list:\n",
    "                NCBI_entry_str_list.remove(\"\")\n",
    "            \n",
    "            # Following the removal of empty strings, the official gene\n",
    "            # symbol is represented by the first list element, but it is\n",
    "            # preceded by the string \"1. \", which encompasses three\n",
    "            # characters\n",
    "            # Hence, prior to comparing the gene names provided by the\n",
    "            # VACV screen data set to the official gene symbols, the\n",
    "            # first list element has to be sliced accordingly\n",
    "            official_gene_symbol = NCBI_entry_str_list[0][3:]\n",
    "            if gene_name != official_gene_symbol:\n",
    "                main_csv_df.at[i, \"Name\"] = official_gene_symbol\n",
    "            \n",
    "            # Irrespective of whether the gene name provided by the VACV\n",
    "            # data set and the official gene symbol match or not, the\n",
    "            # corresponding cell of the \"ID\" feature is populated with\n",
    "            # the NCBI Gene ID harboured by the \"ID_manufacturer\"\n",
    "            # feature (remember that cells of the \"ID\" feature are not\n",
    "            # continuously populated)\n",
    "            main_csv_df.at[i, \"ID\"] = NCBI_Gene_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# As a last step, save the new Pandas DataFrame to a new CSV file\n",
    "main_csv_df.to_csv(\n",
    "    \"Vaccinia_Report_NCBI_Gene_IDs_and_official_gene_symbols_updated.csv\",\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Human-virus PPI', 'Experimental System', 'PubMed ID', 'Virus acronym'], dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# A UniProt reference proteome has been found, which encompasses 219\n",
    "# proteins (the respective web page claims that it encompasses 218\n",
    "# proteins, but a closer look at the FASTA file reveals that they are\n",
    "# 219 instead)\n",
    "# It is checked whether these 219 proteins comprise all interacting VACV\n",
    "# proteins recorded by HVIDB or whether HVIDB contains VACV proteins not\n",
    "# comprised in the reference proteome\n",
    "# First, perform a numerical check\n",
    "HVIDB_VACV_interactions_df = pd.read_csv(\n",
    "    \"all_HVIDB_VACV_interactions.csv\"\n",
    ")\n",
    "print(HVIDB_VACV_interactions_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The VACV UniProt IDs are comprised in the first column bearing the\n",
    "# name \"Human-virus PPI\"; to be more precise, the first column\n",
    "# represents an interaction pair between a human and a VACV protein as a\n",
    "# combination of the respective UniProt IDs, which are separated by a\n",
    "# hyphen\n",
    "# The first UniProt ID represents a human protein, whereas the second\n",
    "# UniProt ID represents a VACV protein\n",
    "interaction_pair_IDs = HVIDB_VACV_interactions_df[\n",
    "    \"Human-virus PPI\"\n",
    "].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "VACV_interaction_IDs = [\n",
    "    pair.split(\"-\")[1] for pair in interaction_pair_IDs\n",
    "]\n",
    "\n",
    "assert len(interaction_pair_IDs) == len(VACV_interaction_IDs), (\n",
    "    \"Something went wrong!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bear in mind that one and the same protein can engage in more than one\n",
    "# PPI, which also applies to VACV\n",
    "# Hence, duplicate UniProt IDs are discarded\n",
    "VACV_unique_interaction_IDs = np.unique(VACV_interaction_IDs).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total amount of VACV proteins involved in interspecies PPI recorded by HVIDB: 61\n"
     ]
    }
   ],
   "source": [
    "# Scrutinising the FASTA file revealed that it indeed comprises 218\n",
    "# sequences as claimed by the UniProt web page (the \">\" character, which\n",
    "# represents the start of a header, has an additional occurrence in an\n",
    "# entry description)\n",
    "assert len(VACV_unique_interaction_IDs) <= 218, (\n",
    "    \"The amount of VACV proteins involved in interspecies PPI recorded \"\n",
    "    \"by HVIDB exceeds the size of the UniProt reference proteome \"\n",
    "    \"(219).\"\n",
    ")\n",
    "\n",
    "print(\n",
    "    \"Total amount of VACV proteins involved in interspecies PPI \"\n",
    "    f\"recorded by HVIDB: {len(VACV_unique_interaction_IDs)}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "18 proteins involved in interspecies PPIs recorded by HVIDB do not occur in the UniProt reference proteome!",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 21\u001b[0m\n\u001b[1;32m     18\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m query_ID \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ref_IDs:\n\u001b[1;32m     19\u001b[0m         absence_list[i] \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[0;32m---> 21\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28msum\u001b[39m(absence_list) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m, (\n\u001b[1;32m     22\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28msum\u001b[39m(absence_list)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m proteins involved in interspecies PPIs \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     23\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrecorded by HVIDB do not occur in the UniProt reference proteome!\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     24\u001b[0m )\n",
      "\u001b[0;31mAssertionError\u001b[0m: 18 proteins involved in interspecies PPIs recorded by HVIDB do not occur in the UniProt reference proteome!"
     ]
    }
   ],
   "source": [
    "# Now, beyond a simple numerical check, it is investigated for each VACV\n",
    "# protein in HVIDB individually whether it is included in the UniProt\n",
    "# reference proteome or not\n",
    "# Conveniently enough, the header contains the respective protein's\n",
    "# UniProt ID, thereby obviating the necessity of ID conversion\n",
    "# Gather the UniProt IDs of the proteins comprised in the UniProt\n",
    "# reference proteome\n",
    "ref_proteome_fasta = fasta.FastaFile.read(\n",
    "    \"uniprotkb_proteome_UP000000344_VACV_WR_04_11_2024.fasta\"\n",
    ")\n",
    "ref_proteome_dict = fasta.get_sequences(ref_proteome_fasta)\n",
    "ref_proteome_headers = list(ref_proteome_dict.keys())\n",
    "ref_IDs = [header.split(\"|\")[1] for header in ref_proteome_headers]\n",
    "\n",
    "absence_list = [0] * len(VACV_unique_interaction_IDs)\n",
    "\n",
    "for i, query_ID in enumerate(VACV_unique_interaction_IDs):\n",
    "    if query_ID not in ref_IDs:\n",
    "        absence_list[i] = 1\n",
    "\n",
    "assert sum(absence_list) == 0, (\n",
    "    f\"{sum(absence_list)} proteins involved in interspecies PPIs \"\n",
    "    \"recorded by HVIDB do not occur in the UniProt reference proteome!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determine the UniProt IDs of those 18 proteins not comprised in the\n",
    "# reference proteome\n",
    "outcast_protein_indices = np.nonzero(absence_list)\n",
    "outcast_protein_IDs = np.array(VACV_unique_interaction_IDs)[\n",
    "    outcast_protein_indices\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A4GDF4' 'O57173' 'O57263' 'P04299' 'P20505' 'P20639' 'P68318' 'P68451'\n",
      " 'P68467' 'Q1M1E0' 'Q49PX0' 'Q49QD4' 'Q711A3' 'Q76RC6' 'Q76ZR2' 'Q86638'\n",
      " 'Q8BDF8' 'Q8UYL3']\n"
     ]
    }
   ],
   "source": [
    "print(outcast_protein_IDs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Upon closer scrutiny, it becomes apparent that while one \"outcast\"\n",
    "# protein ID (P04299) belongs to VACV strain Western Reserve and is not\n",
    "# included in the reference genome, the remaining ones belong to either\n",
    "# VACV in general or other strains, such as VACV strain Ankara or VACV\n",
    "# strain Copenhagen\n",
    "# As a consequence, two measures are taken, the first of which consists\n",
    "# of discarding the VACV reference proteome in favour of the entirety of\n",
    "# VACV proteins available on UniProt (440 in total as of 5th November\n",
    "# 2024) and the second of which involves discarding the interspecies\n",
    "# PPIs not specific to VACV strain Western Reserve from the HVIDB data\n",
    "# set\n",
    "# The aforementioned 440 VACV proteines are downloaded from UniProt in\n",
    "# the FASTA format including both canonical sequences and isoforms (i.e.\n",
    "# \"FASTA (canonical & isoform)\" has been chosen as \"Format\" option)\n",
    "uniprot_IDs_to_exclude = [\n",
    "    'A4GDF4', 'O57173', 'O57263', 'P04299', 'P20505', 'P20639',\n",
    "    'P68318', 'P68451', 'P68467', 'Q1M1E0', 'Q49PX0', 'Q49QD4',\n",
    "    'Q711A3', 'Q76RC6', 'Q76ZR2', 'Q86638', 'Q8BDF8', 'Q8UYL3'\n",
    "]\n",
    "# Determine the amount of interspecies interactions the outcast UniProt\n",
    "# IDs are involved in in order to perform a sanity check later on\n",
    "n_outcast_interactions = sum([\n",
    "    int_pair.split(\"-\")[1] in uniprot_IDs_to_exclude\n",
    "    for int_pair in HVIDB_VACV_interactions_df[\"Human-virus PPI\"]\n",
    "])\n",
    "\n",
    "HVIDB_VACV_WR_interactions_df = HVIDB_VACV_interactions_df[\n",
    "    ~np.array([\n",
    "        int_pair.split(\"-\")[1] in uniprot_IDs_to_exclude\n",
    "        for int_pair in HVIDB_VACV_interactions_df[\"Human-virus PPI\"]\n",
    "    ])\n",
    "]\n",
    "\n",
    "assert (\n",
    "    len(HVIDB_VACV_WR_interactions_df) == (\n",
    "        len(HVIDB_VACV_interactions_df) - n_outcast_interactions\n",
    "    )\n",
    "), (\n",
    "    \"Something went wrong while filtering out interspecies \"\n",
    "    \"interactions involving the outcast UniProt IDs!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the DataFrame containing exclusively VACV proteins specific to\n",
    "# the Western Reserve strain to a CSV file\n",
    "HVIDB_VACV_WR_interactions_df.to_csv(\n",
    "    \"all_HVIDB_VACV_WR_interactions.csv\",\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total amount of interspecies PPIs specific to VACV strain Western Reserve is 415.\n"
     ]
    }
   ],
   "source": [
    "# Output the total amount of interspecies PPIs specific to VACV strain\n",
    "# Western Reserve\n",
    "print(\n",
    "    \"The total amount of interspecies PPIs specific to VACV strain \"\n",
    "    f\"Western Reserve is {len(HVIDB_VACV_WR_interactions_df)}.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a FASTA file containing all VACV WR proteins occurring in the\n",
    "# HVIDB data set\n",
    "# This is done by first retrieving the FASTA entries of the individual\n",
    "# proteins from UniProt and subsequently adding them to an empty FASTA\n",
    "# file\n",
    "VACV_WR_unique_prots = np.unique(\n",
    "    [\n",
    "        int_pair.split(\"-\")[1]\n",
    "        for int_pair in HVIDB_VACV_WR_interactions_df[\"Human-virus PPI\"]\n",
    "    ]\n",
    ").tolist()\n",
    "\n",
    "uniprot_entries_list = uniprot.fetch(\n",
    "    ids=VACV_WR_unique_prots,\n",
    "    format=\"fasta\"\n",
    ")\n",
    "\n",
    "VACV_WR_prots_fasta_file = fasta.FastaFile()\n",
    "\n",
    "for io_object in uniprot_entries_list:\n",
    "    # As a list of IDs is passed to \"uniprot.fetch()\" and no target path\n",
    "    # is specified, a list of StringIO objects is returned which can be\n",
    "    # iterated over\n",
    "    current_fasta_file = fasta.FastaFile.read(io_object)\n",
    "    # Each file contains only one entry; hence, the first and only\n",
    "    # element is retrieved from the iterator returned by the `items()`\n",
    "    # method\n",
    "    header, seq_str = list(current_fasta_file.items())[0]\n",
    "\n",
    "    # Finally, append the entry to the FASTA file\n",
    "    VACV_WR_prots_fasta_file[header] = seq_str\n",
    "\n",
    "assert len(VACV_WR_prots_fasta_file) == len(VACV_WR_unique_prots), (\n",
    "    \"Something went wrong while creating the FASTA file for the VACV \"\n",
    "    \"strain Western Reserve proteins!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the FASTA file harbouring all VACV strain Western Reserve\n",
    "# proteins to disk\n",
    "VACV_WR_prots_fasta_file.write(\"VACV_WR_prots_in_HVIDB.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In retrospect, it probably would have been safer to just look the\n",
    "# sequences up in the FASTA file encompassing all VACV strain Western\n",
    "# Reserve (VACV WR) proteins (sometimes, when querying databases,\n",
    "# incomplete information transmission may occur)\n",
    "# Therefore, it is investigated whether complete information\n",
    "# transmission took place by comparing the sequences obtained by\n",
    "# database quering to the ones deposited in the FASTA\n",
    "all_VACV_WR_prots_fasta = fasta.FastaFile.read(\n",
    "    \"uniprotkb_taxonomy_id_10254_all_VACV_WR_prots_05_11_2024.fasta\"\n",
    ")\n",
    "all_VACV_WR_prots_dict = fasta.get_sequences(all_VACV_WR_prots_fasta)\n",
    "\n",
    "correct_seqs_list = [False] * len(VACV_WR_prots_fasta_file)\n",
    "\n",
    "for i, (header, seq_str) in enumerate(VACV_WR_prots_fasta_file.items()):\n",
    "    try:\n",
    "        ref_seq = str(all_VACV_WR_prots_dict[header])\n",
    "    except KeyError:\n",
    "        print(\"One header was incompletely transmitted!\")\n",
    "        continue\n",
    "    \n",
    "    if seq_str == ref_seq:\n",
    "        correct_seqs_list[i] = True\n",
    "\n",
    "assert all(correct_seqs_list), (\n",
    "    \"Complete information transmission did not occur for all proteins!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "HVIDB_VACV_WR_interactions_df = pd.read_csv(\n",
    "    \"all_HVIDB_VACV_WR_interactions.csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repeat the procedure for the human counterpart, i.e. the human\n",
    "# proteins involved in interspecies PPIs recorded by HVIDB\n",
    "human_unique_prots_in_HVIDB = np.unique(\n",
    "    [\n",
    "        int_pair.split(\"-\")[0]\n",
    "        for int_pair in HVIDB_VACV_WR_interactions_df[\"Human-virus PPI\"]\n",
    "    ]\n",
    ").tolist()\n",
    "\n",
    "# As the total amount of human proteins is way too large to be processed\n",
    "# in one single database query (357), the corresponding list is split\n",
    "# into chunks each encompassing 20 UniProt IDs\n",
    "CHUNK_SIZE = 20\n",
    "n_chunks = math.ceil(len(human_unique_prots_in_HVIDB) / CHUNK_SIZE)\n",
    "human_prots_chunks = [\n",
    "    human_unique_prots_in_HVIDB[i * CHUNK_SIZE : (i + 1) * CHUNK_SIZE]\n",
    "    for i in range(n_chunks)\n",
    "]\n",
    "\n",
    "human_prots_in_HVIDB_fasta_file = fasta.FastaFile()\n",
    "\n",
    "for i, chunk in enumerate(human_prots_chunks):\n",
    "    # # Introducing kind of a numerical progress bar\n",
    "    # if i % 20 == 0:\n",
    "    #     print(i)\n",
    "    time.sleep(2)\n",
    "\n",
    "    uniprot_query_unsuccessful = True\n",
    "    while uniprot_query_unsuccessful:\n",
    "        try:\n",
    "            uniprot_entries_list = uniprot.fetch(\n",
    "                ids=chunk,\n",
    "                format=\"fasta\"\n",
    "            )\n",
    "\n",
    "            for io_object in uniprot_entries_list:\n",
    "                try:\n",
    "                    current_fasta_file = fasta.FastaFile.read(io_object)\n",
    "                    header, seq_str = list(current_fasta_file.items())[0]\n",
    "                    human_prots_in_HVIDB_fasta_file[header] = seq_str\n",
    "                except:\n",
    "                    continue\n",
    "            uniprot_query_unsuccessful = False\n",
    "        except:\n",
    "            time.sleep(1)\n",
    "\n",
    "human_prots_in_HVIDB_fasta_file.write(\n",
    "    \"human_prots_in_HVIDB_VACV_WR_data_set.fasta\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the following 3 UniProt IDs, no UniProt entry has been retrieved: ['A0A087X117', 'A0A0A0MR88', 'A0A0A0MTS7', 'A0A3B3IS20', 'H0Y4G9']\n"
     ]
    }
   ],
   "source": [
    "# Investigate whether a UniProt entry has been retrieved for each and\n",
    "# every ID\n",
    "if len(human_prots_in_HVIDB_fasta_file) != len(human_unique_prots_in_HVIDB):\n",
    "    human_prots_headers = human_prots_in_HVIDB_fasta_file.keys()\n",
    "    uniprot_ids_in_fasta_file = [\n",
    "        header.split(\"|\")[1] for header in human_prots_headers\n",
    "    ]\n",
    "\n",
    "    n_prots_not_included = (\n",
    "        len(human_unique_prots_in_HVIDB)\n",
    "        -\n",
    "        len(human_prots_in_HVIDB_fasta_file)\n",
    "    )\n",
    "    prots_not_included = []\n",
    "\n",
    "    for uniprot_id in human_unique_prots_in_HVIDB:\n",
    "        if uniprot_id not in uniprot_ids_in_fasta_file:\n",
    "            prots_not_included.append(uniprot_id)\n",
    "    \n",
    "    print(\n",
    "        f\"For the following {n_prots_not_included} UniProt IDs, no \"\n",
    "        f\"UniProt entry has been retrieved: {prots_not_included}\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of HVIDB VACV WR-human PPI data set prior to filtering out IDs removed from UniProt: 415\n",
      "Length of HVIDB VACV WR-human PPI data set after filtering out IDs removed from UniProt: 412\n"
     ]
    }
   ],
   "source": [
    "# Surprisingly, although the FASTA file has only three proteins less\n",
    "# than there are unique human proteins in the HVIDB data set, five\n",
    "# UniProt IDs are found not to occur in the FASTA file\n",
    "# This is subjected to further scrutiny\n",
    "#\n",
    "# A0A087X117: has indeed been removed from UniProt\n",
    "#\n",
    "# A0A0A0MR88: has indeed been removed from UniProt\n",
    "#\n",
    "# A0A0A0MTS7: the entry still exists, but has been merged into C9JQJ2\n",
    "#\n",
    "# A0A3B3IS20: the entry still exists, but has been merged into\n",
    "# A0A3B3IRW5\n",
    "#\n",
    "# H0Y4G9: has indeed been removed from UniProt\n",
    "#\n",
    "# The DataFrame harbouring the HVIDB PPIs is updated to reflect these\n",
    "# changes, following which the updated DataFrame is saved to disk\n",
    "len_prior_to_filtering = len(HVIDB_VACV_WR_interactions_df)\n",
    "print(\n",
    "    \"Length of HVIDB VACV WR-human PPI data set prior to filtering out \"\n",
    "    f\"IDs removed from UniProt: {len_prior_to_filtering}\"\n",
    ")\n",
    "\n",
    "HVIDB_VACV_WR_interactions_df = HVIDB_VACV_WR_interactions_df[\n",
    "    ~(\n",
    "        HVIDB_VACV_WR_interactions_df[\"Human-virus PPI\"].str.contains(\n",
    "            \"A0A087X117\"\n",
    "        )\n",
    "        |\n",
    "        HVIDB_VACV_WR_interactions_df[\"Human-virus PPI\"].str.contains(\n",
    "            \"A0A0A0MR88\"\n",
    "        )\n",
    "        |\n",
    "        HVIDB_VACV_WR_interactions_df[\"Human-virus PPI\"].str.contains(\n",
    "            \"H0Y4G9\"\n",
    "        )\n",
    "    )\n",
    "]\n",
    "\n",
    "assert len(HVIDB_VACV_WR_interactions_df) == (\n",
    "    len_prior_to_filtering - 3\n",
    "), (\n",
    "    \"Something went wrong while filtering out interactions involving \"\n",
    "    \"proteins removed from UniProt!\"\n",
    ")\n",
    "\n",
    "print(\n",
    "    \"Length of HVIDB VACV WR-human PPI data set after filtering out \"\n",
    "    f\"IDs removed from UniProt: {len(HVIDB_VACV_WR_interactions_df)}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, update the IDs that have been merged into other ones\n",
    "# To this end, iterate over the DataFrame rows\n",
    "for i, row in HVIDB_VACV_WR_interactions_df.iterrows():\n",
    "    row_human_prot_id = row[\"Human-virus PPI\"].split(\"-\")[0]\n",
    "    row_virus_prot_id = row[\"Human-virus PPI\"].split(\"-\")[1]\n",
    "    if row_human_prot_id == \"A0A0A0MTS7\":\n",
    "        new_interaction_pair = \"C9JQJ2\" + \"-\" + row_virus_prot_id\n",
    "        # Bear in mind that `.iterrows()` returns the rows' index, which\n",
    "        # is not necessarily identical to the respective integer\n",
    "        # positions\n",
    "        # Hence, employing `.at()` is preferred to using `.iat()`\n",
    "        HVIDB_VACV_WR_interactions_df.at[\n",
    "            i, \"Human-virus PPI\"\n",
    "        ] = new_interaction_pair\n",
    "    elif row_human_prot_id == \"A0A3B3IS20\":\n",
    "        new_interaction_pair = \"A0A3B3IRW5\" + \"-\" + row_virus_prot_id\n",
    "        HVIDB_VACV_WR_interactions_df.at[\n",
    "            i, \"Human-virus PPI\"\n",
    "        ] = new_interaction_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finally, save the updated DataFrame to Disk\n",
    "HVIDB_VACV_WR_interactions_df.to_csv(\n",
    "    \"all_HVIDB_VACV_WR_interactions.csv\",\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform a final sanity check\n",
    "list_of_human_prots_in_HVIDB_fasta_file = np.unique([\n",
    "    int_pair.split(\"-\")[0]\n",
    "    for int_pair in HVIDB_VACV_WR_interactions_df[\"Human-virus PPI\"]\n",
    "]).tolist()\n",
    "\n",
    "ids_not_supposed_to_occur = [\n",
    "    \"A0A087X117\",\n",
    "    \"A0A0A0MR88\",\n",
    "    \"A0A0A0MTS7\",\n",
    "    \"A0A3B3IS20\",\n",
    "    \"H0Y4G9\"\n",
    "]\n",
    "\n",
    "assert all([\n",
    "    outcast_id not in list_of_human_prots_in_HVIDB_fasta_file\n",
    "    for outcast_id in ids_not_supposed_to_occur\n",
    "]), (\n",
    "    \"Something went wrong while removing invalid IDs and replacing IDs \"\n",
    "    \"that were merged into others!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/homebrew/Caskroom/miniforge/base/envs/bioinformatics/lib/python3.10/site-packages/biotite/sequence/io/fasta/convert.py:253: UserWarning: ProteinSequence objects do not support selenocysteine (U), occurrences were substituted by cysteine (C)\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n",
      "One header was incompletely transmitted!\n"
     ]
    },
    {
     "ename": "AssertionError",
     "evalue": "Complete information transmission did not occur for all proteins!",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[28], line 30\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m seq_str \u001b[38;5;241m==\u001b[39m ref_seq:\n\u001b[1;32m     28\u001b[0m         correct_seqs_list[i] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m---> 30\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mall\u001b[39m(correct_seqs_list), (\n\u001b[1;32m     31\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mComplete information transmission did not occur for all proteins!\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     32\u001b[0m )\n",
      "\u001b[0;31mAssertionError\u001b[0m: Complete information transmission did not occur for all proteins!"
     ]
    }
   ],
   "source": [
    "# Now, complete information transmission is also checked for the human\n",
    "# interaction partners\n",
    "# To this end, the reference proteome of Homo sapiens is downloaded and\n",
    "# loaded into a DataFrame\n",
    "homo_sapiens_ref_proteome_fasta = fasta.FastaFile.read(\n",
    "    \"uniprotkb_Homo_sapiens_reference_proteome_06_11_2024.fasta\"\n",
    ")\n",
    "homo_sapiens_ref_proteome_dict = fasta.get_sequences(\n",
    "    homo_sapiens_ref_proteome_fasta\n",
    ")\n",
    "\n",
    "human_prots_in_HVIDB_fasta_file = fasta.FastaFile.read(\n",
    "    \"human_prots_in_HVIDB_VACV_WR_data_set.fasta\"\n",
    ")\n",
    "\n",
    "correct_seqs_list = [False] * len(human_prots_in_HVIDB_fasta_file)\n",
    "\n",
    "for i, (header, seq_str) in enumerate(\n",
    "    human_prots_in_HVIDB_fasta_file.items()\n",
    "):\n",
    "    try:\n",
    "        ref_seq = str(homo_sapiens_ref_proteome_dict[header])\n",
    "    except KeyError:\n",
    "        print(\"One header was incompletely transmitted!\")\n",
    "        continue\n",
    "\n",
    "    if seq_str == ref_seq:\n",
    "        correct_seqs_list[i] = True\n",
    "\n",
    "assert all(correct_seqs_list), (\n",
    "    \"Complete information transmission did not occur for all proteins!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A0N0N7' 'A0N0Q3' 'A3KPC7' 'A4FTV9' 'A8ASI8' 'B2R4P9' 'B2R4R0' 'B2R4S9'\n",
      " 'B2RDW1' 'B2ZZ89' 'B4DJ51' 'D9YZV4' 'D9ZGF2' 'E5KTA5' 'E9KL37' 'F4ZW62'\n",
      " 'Q548T7']\n"
     ]
    }
   ],
   "source": [
    "# The discrepancy of 17 proteins is entirely attributable to KeyErrors,\n",
    "# i.e. the fact that the respective headers are not contained in the\n",
    "# dictionary of the reference proteome\n",
    "# Determine the UniProt IDs of those 17 proteins via Boolean indexing\n",
    "human_prots_in_HVIDB_headers = list(\n",
    "    human_prots_in_HVIDB_fasta_file.keys()\n",
    ")\n",
    "human_prots_in_HVIDB_IDs = [\n",
    "    header.split(\"|\")[1] for header in human_prots_in_HVIDB_headers\n",
    "]\n",
    "\n",
    "discrepancy_ids = np.array(\n",
    "    human_prots_in_HVIDB_IDs\n",
    ")[~np.array(correct_seqs_list)]\n",
    "\n",
    "print(discrepancy_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Querying the UniProt database with the 17 IDs causing the discrepancy\n",
    "# revealed both that they exist and there ID has not been changed\n",
    "# Thus, they are simply not included in the reference proteome of Homo\n",
    "# sapiens\n",
    "# Complete information transmission is nevertheless checked by verifying\n",
    "# that the sequences have the expected length\n",
    "discrepancy_seq_lengths = [\n",
    "    397, 127, 128, 130, 195, 136, 103, 126, 156, 2364, 149, 284, 3177,\n",
    "    313, 1012, 390, 677\n",
    "]\n",
    "\n",
    "discrepancy_headers = np.array(human_prots_in_HVIDB_headers)[\n",
    "    ~np.array(correct_seqs_list)\n",
    "]\n",
    "\n",
    "human_prots_in_HVIDB_dict = fasta.get_sequences(\n",
    "    human_prots_in_HVIDB_fasta_file\n",
    ")\n",
    "\n",
    "correct_lengths_list = []\n",
    "\n",
    "for header, ref_len in zip(discrepancy_headers, discrepancy_seq_lengths):\n",
    "    transmitted_seq = str(human_prots_in_HVIDB_dict[header])\n",
    "    correct_lengths_list.append(\n",
    "        len(transmitted_seq) == ref_len\n",
    "    )\n",
    "\n",
    "assert all(correct_lengths_list), (\n",
    "    f\"For {len(correct_lengths_list) - sum(correct_lengths_list)} of \"\n",
    "    \"the 17 IDs causing the discrepancy, the respective sequence has \"\n",
    "    \"not been correctly transmitted.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I stopped here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Code I cannot remember the purpose of underneath...\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "VACV_proteome_fasta = ref_proteome_fasta.copy()\n",
    "\n",
    "# Querying the UniProtKB database with a couple of the IDs not included\n",
    "# in the reference genome reveals that some IDs are unreviewed, while\n",
    "# others have a low annotation score and where therefore not\n",
    "# incorporated\n",
    "# The functionalities provided by Biotite are again used in order to\n",
    "# retrieve the respective proteins' sequences from the UniProt database,\n",
    "# add them to the sequences from the reference proteome and save the\n",
    "# resulting concatenation to a new FASTA file\n",
    "# As a list of IDs is passed to `uniprot.fetch()` and `target_path` is\n",
    "# not specified, a list of StringIO objects is returned which can be\n",
    "# iterated over\n",
    "for io_object in uniprot.fetch(outcast_protein_IDs, \"fasta\"):\n",
    "    # Read the StringIO object into a Fasta file\n",
    "    fasta_file = fasta.FastaFile.read(io_object)\n",
    "    # Each file contains only one entry; hence, the first and only\n",
    "    # element is retrieved from the the iterator returned by the \n",
    "    # `items()` method\n",
    "    header, seq_str = list(fasta_file.items())[0]\n",
    "    \n",
    "    VACV_proteome_fasta[header] = seq_str\n",
    "\n",
    "# Strangely enough, creating a deep copy of a FASTA file resets its\n",
    "# length to zero although it is not empty (applies to Biotite version\n",
    "# 0.39.0); hence, instead of evaluating whether the new file's length\n",
    "# equals the old file's length plus the amount of new entries, it is\n",
    "# checked whether the new file's length equals the amount of newly added\n",
    "# items\n",
    "assert len(VACV_proteome_fasta) == 18, (\n",
    "    \"Something went wrong while adding the 18 entries to the reference \"\n",
    "    \"proteome!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remember that the UniProtKB (KB stands for knowledgebase) is composed\n",
    "# of two sections, which are UniProtKB/Swiss-Prot and UniProtKB/TrEMBL\n",
    "# The former is characterised by the fact that it contains exclusively\n",
    "# reviewed, i.e. manually curated entries\n",
    "# The latter, however, are unreviewed\n",
    "# In order to incorporate both sections of UniProtKB, \"UniProtKB\" is\n",
    "# chosen as the destination database (\"To database\")\n",
    "# As source database (\"from database\"),\n",
    "# \"Gene annotation databases/Gene ID\" is chosen\n",
    "# Obtain the unique gene IDs\n",
    "csv_df = pd.read_csv(\n",
    "    \"Vaccinia_Report_updated_and_errors_fixed.csv\",\n",
    "    sep=\"\\t\",\n",
    "    dtype=dtype_dict\n",
    ")\n",
    "\n",
    "# Using those settings, \n",
    "\n",
    "# Using the parameters \"from: UniProt/Gene Name\",\n",
    "# \"to: UniProtKB-Swiss-Prot\" and \"taxId: 9606\", 18,185 of the 18,396\n",
    "# human protein IDs were mapped to 18,645 results, whereas 213 ID were\n",
    "# not mapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A0A075B749\n",
      "A0A087X117\n",
      "A0A0A0MR88\n",
      "A0A0A0MTS7\n",
      "A0A0D9SG04\n",
      "A0A0J9YX62\n",
      "A0A0U1RRM6\n",
      "A0A1B0GTL5\n",
      "A0A2R8Y5A3\n",
      "A0A3B3IS20\n",
      "A0A3B3ISQ4\n",
      "A0N0N7\n",
      "A0N0Q3\n",
      "A3KPC7\n",
      "A4FTV9\n",
      "A6NFX8\n",
      "A6NNZ2\n",
      "A7MCY6\n",
      "A8ASI8\n",
      "A8MUS3\n",
      "B2R4P9\n",
      "B2R4R0\n",
      "B2R4S9\n",
      "B2RDW1\n",
      "B2ZZ89\n",
      "B4DJ51\n",
      "B4DLJ1\n",
      "B8ZZN6\n",
      "D9YZV4\n",
      "D9ZGF2\n",
      "E5KTA5\n",
      "E9KL37\n",
      "E9PDI4\n",
      "F4ZW62\n",
      "F8VVA7\n",
      "F8VXC8\n",
      "F8VZQ9\n",
      "F8WBV6\n",
      "G3V5R9\n",
      "G5E9I4\n",
      "H0Y4G9\n",
      "H3BSR6\n",
      "J3QK89\n",
      "K7EQ78\n",
      "K7ERV3\n",
      "O00151\n",
      "O00206\n",
      "O00327\n",
      "O00339\n",
      "O00391\n",
      "O00459\n",
      "O00499\n",
      "O00515\n",
      "O00571\n",
      "O00585\n",
      "O14578\n",
      "O14654\n",
      "O14733\n",
      "O14920\n",
      "O14933\n",
      "O15020\n",
      "O15021\n",
      "O15037\n",
      "O15111\n",
      "O15131\n",
      "O15444\n",
      "O43187\n",
      "O43390\n",
      "O43521\n",
      "O43815\n",
      "O60506\n",
      "O60814\n",
      "O75113\n",
      "O75152\n",
      "O75190\n",
      "O75376\n",
      "O75398\n",
      "O75477\n",
      "O75665\n",
      "O94763\n",
      "O95400\n",
      "O95793\n",
      "P00451\n",
      "P01375\n",
      "P01579\n",
      "P01903\n",
      "P02511\n",
      "P02775\n",
      "P02776\n",
      "P04083\n",
      "P04183\n",
      "P04439\n",
      "P04792\n",
      "P04908\n",
      "P05023\n",
      "P05141\n",
      "P05166\n",
      "P05412\n",
      "P06733\n",
      "P06748\n",
      "P07355\n",
      "P07437\n",
      "P07910\n",
      "P08238\n",
      "P09104\n",
      "P09493\n",
      "P09651\n",
      "P0C0S8\n",
      "P0CG48\n",
      "P0DP23\n",
      "P11021\n",
      "P12036\n",
      "P12111\n",
      "P12235\n",
      "P12236\n",
      "P12273\n",
      "P12277\n",
      "P13051\n",
      "P13637\n",
      "P13861\n",
      "P13929\n",
      "P14923\n",
      "P15924\n",
      "P18583\n",
      "P19013\n",
      "P19237\n",
      "P19525\n",
      "P19876\n",
      "P20042\n",
      "P20671\n",
      "P20700\n",
      "P22362\n",
      "P22492\n",
      "P22531\n",
      "P22532\n",
      "P22626\n",
      "P23025\n",
      "P23246\n",
      "P24001\n",
      "P24539\n",
      "P25391\n",
      "P27449\n",
      "P27986\n",
      "P28562\n",
      "P28838\n",
      "P31025\n",
      "P31151\n",
      "P33176\n",
      "P35221\n",
      "P35222\n",
      "P35268\n",
      "P35325\n",
      "P35326\n",
      "P38159\n",
      "P42224\n",
      "P42336\n",
      "P42338\n",
      "P43631\n",
      "P46776\n",
      "P46778\n",
      "P46939\n",
      "P48788\n",
      "P49792\n",
      "P50552\n",
      "P52179\n",
      "P52292\n",
      "P54652\n",
      "P55265\n",
      "P55327\n",
      "P55957\n",
      "P57053\n",
      "P57738\n",
      "P58753\n",
      "P58876\n",
      "P61923\n",
      "P61978\n",
      "P62266\n",
      "P62750\n",
      "P62805\n",
      "P62807\n",
      "P62979\n",
      "P63010\n",
      "P63165\n",
      "P63208\n",
      "P63244\n",
      "P67809\n",
      "P68104\n",
      "P68371\n",
      "P68431\n",
      "P78344\n",
      "P78423\n",
      "P82979\n",
      "P84101\n",
      "P84243\n",
      "P98179\n",
      "Q00325\n",
      "Q00610\n",
      "Q00839\n",
      "Q01082\n",
      "Q01850\n",
      "Q01955\n",
      "Q02539\n",
      "Q03252\n",
      "Q05086\n",
      "Q05639\n",
      "Q05BQ5\n",
      "Q06481\n",
      "Q07666\n",
      "Q07812\n",
      "Q07866\n",
      "Q08211\n",
      "Q08379\n",
      "Q08AE8\n",
      "Q12905\n",
      "Q12906\n",
      "Q13085\n",
      "Q13162\n",
      "Q13227\n",
      "Q13283\n",
      "Q13427\n",
      "Q13618\n",
      "Q13753\n",
      "Q13794\n",
      "Q13885\n",
      "Q14103\n",
      "Q14141\n",
      "Q14160\n",
      "Q14444\n",
      "Q15057\n",
      "Q15084\n",
      "Q15155\n",
      "Q15643\n",
      "Q15797\n",
      "Q15853\n",
      "Q16543\n",
      "Q16611\n",
      "Q16695\n",
      "Q16777\n",
      "Q2TBE0\n",
      "Q3ZCM7\n",
      "Q4G0J3\n",
      "Q53H80\n",
      "Q53SF7\n",
      "Q548T7\n",
      "Q5EBL4\n",
      "Q5JRA6\n",
      "Q5QNW6\n",
      "Q5VTE0\n",
      "Q5VU43\n",
      "Q66GS9\n",
      "Q6FI13\n",
      "Q6P0Q8\n",
      "Q6P597\n",
      "Q6P5S2\n",
      "Q6UWP8\n",
      "Q6ZN40\n",
      "Q71DI3\n",
      "Q71U36\n",
      "Q76M96\n",
      "Q7L7L0\n",
      "Q7Z406\n",
      "Q7Z7A1\n",
      "Q86VP6\n",
      "Q86X95\n",
      "Q86YZ3\n",
      "Q8IUC6\n",
      "Q8IUR0\n",
      "Q8IVF2\n",
      "Q8IWX8\n",
      "Q8IXT5\n",
      "Q8IY57\n",
      "Q8N8S7\n",
      "Q8NFZ5\n",
      "Q8TAQ2\n",
      "Q8TDL5\n",
      "Q8WTS1\n",
      "Q8WUM4\n",
      "Q8WV24\n",
      "Q8WW38\n",
      "Q8WWI1\n",
      "Q8WXH0\n",
      "Q8WZ42\n",
      "Q92187\n",
      "Q92560\n",
      "Q92569\n",
      "Q92583\n",
      "Q92599\n",
      "Q92624\n",
      "Q92844\n",
      "Q92878\n",
      "Q92934\n",
      "Q93008\n",
      "Q93077\n",
      "Q93079\n",
      "Q93086\n",
      "Q96AH0\n",
      "Q96C34\n",
      "Q96DA0\n",
      "Q96E22\n",
      "Q96ES6\n",
      "Q96F44\n",
      "Q96JA1\n",
      "Q96JH7\n",
      "Q96KK5\n",
      "Q96PG8\n",
      "Q96RQ9\n",
      "Q99661\n",
      "Q99675\n",
      "Q99729\n",
      "Q99836\n",
      "Q99877\n",
      "Q99878\n",
      "Q99879\n",
      "Q99880\n",
      "Q9BQE3\n",
      "Q9BQS8\n",
      "Q9BRK4\n",
      "Q9BS26\n",
      "Q9BT92\n",
      "Q9BTM1\n",
      "Q9BVA1\n",
      "Q9BXF6\n",
      "Q9BXH1\n",
      "Q9BYE4\n",
      "Q9BYP7\n",
      "Q9C0F3\n",
      "Q9GZT9\n",
      "Q9H2G2\n",
      "Q9H2G4\n",
      "Q9H334\n",
      "Q9H4A3\n",
      "Q9H6S1\n",
      "Q9H6S3\n",
      "Q9HCC0\n",
      "Q9HCU9\n",
      "Q9NP74\n",
      "Q9NP98\n",
      "Q9NQ29\n",
      "Q9NRJ3\n",
      "Q9NRQ2\n",
      "Q9NSA3\n",
      "Q9NUL3\n",
      "Q9NUU7\n",
      "Q9P0K7\n",
      "Q9P2P1\n",
      "Q9UBB5\n",
      "Q9UH99\n",
      "Q9UHD2\n",
      "Q9UKK9\n",
      "Q9UMR2\n",
      "Q9UPT5\n",
      "Q9UQF2\n",
      "Q9Y258\n",
      "Q9Y2H9\n",
      "Q9Y3B3\n",
      "Q9Y3C5\n",
      "Q9Y3S1\n",
      "Q9Y3T6\n",
      "Q9Y4E1\n",
      "Q9Y4K3\n",
      "Q9Y4Z0\n",
      "Q9Y5Q9\n",
      "Q9Y5S2\n",
      "Q9Y5X3\n",
      "Q9Y618\n",
      "Q9Y679\n",
      "Q9Y6K9\n",
      "V9GZ56\n"
     ]
    }
   ],
   "source": [
    "VACV_interactions_df = pd.read_csv(\n",
    "    \"all_HVIDB_VACV_interactions.csv\"\n",
    ")\n",
    "\n",
    "interaction_pairs = VACV_interactions_df[\"Human-virus PPI\"].to_list()\n",
    "# Bear in mind that in the interaction pairs, the first ID preceding the\n",
    "# hyphen represents the human protein, whereas the second ID following\n",
    "# the hyphen represents the virus protein\n",
    "human_uniprot_ids = [pair.split(\"-\")[0] for pair in interaction_pairs]\n",
    "unique_human_uniprot_ids = np.unique(human_uniprot_ids)\n",
    "\n",
    "for uniprot_id in unique_human_uniprot_ids:\n",
    "    print(uniprot_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Regarding the mapping of UniProt IDs of human interaction partners to\n",
    "# gene names, some IDs seem to have been mapped to multiple names\n",
    "# This is subjected to further scrutiny\n",
    "human_mapping_df = pd.read_csv(\n",
    "    \"HVIDB_human_interaction_partner_mapping_to_gene_names.tsv\",\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "\n",
    "unique_uniprot_ids = human_mapping_df[\"From\"].unique()\n",
    "\n",
    "multiple_mappings_list = []\n",
    "for uniprot_id in unique_uniprot_ids:\n",
    "    if (human_mapping_df[\"From\"] == uniprot_id).sum() > 1:\n",
    "        multiple_mappings_list.append(uniprot_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P04908\n",
      "P0C0S8\n",
      "P62805\n",
      "P62807\n",
      "P68431\n",
      "P84243\n",
      "Q6FI13\n",
      "Q71DI3\n"
     ]
    }
   ],
   "source": [
    "for uniprot_id in multiple_mappings_list:\n",
    "    print(uniprot_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinformatics",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
