{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nThe purpose of this Jupyter notebook is to preprocess/prepare the VACV\\nscreen TSV file for the evaluation of off-target filtering performance\\n(i.e. for populating table number 2).\\n'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "The purpose of this Jupyter notebook is to preprocess/prepare the VACV\n",
    "screen TSV file for the evaluation of off-target filtering performance\n",
    "(i.e. for populating table number 2).\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/j1/64kbg_f11z97kx1dw__420vh0000gn/T/ipykernel_1310/388131955.py:9: DtypeWarning: Columns (42,74) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  VACV_screen_df = pd.read_csv(\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "path_to_VACV_screen_tsv_file = (\n",
    "    \"/Users/jacobanter/Documents/Code/VACV_screen/VACV_Report_only_\"\n",
    "    \"valid_single_pooled_siRNA_and_esiRNA_single_entries_only_without_\"\n",
    "    \"Qiagen_mismatches.tsv\"\n",
    ")\n",
    "\n",
    "VACV_screen_df = pd.read_csv(\n",
    "    path_to_VACV_screen_tsv_file,\n",
    "    sep=\"\\t\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# As the evaluation of off-target filtering performance is done on the\n",
    "# same data set as in the comparison of protein language models, the\n",
    "# corresponding data set (confirmed positive interactions and reliable\n",
    "# negative interactions between human and VACV WR proteins) is loaded as\n",
    "# well\n",
    "path_to_combined_VACV_WR_human_PPI_data_set = (\n",
    "    \"/Users/jacobanter/Documents/Code/VACV_screen/HVIDB_pos_instances_\"\n",
    "    \"with_nucleolus_neg_instances/VACV_WR_pos_and_nucleolus_prots_\"\n",
    "    \"neg_PPI_instances.tsv\"\n",
    ")\n",
    "\n",
    "VACV_WR_pos_and_neg_PPIs_df = pd.read_csv(\n",
    "    path_to_combined_VACV_WR_human_PPI_data_set,\n",
    "    sep=\"\\t\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "Not all human interaction partners comprised in the PPI data set are interrogated in the VACV screen!",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 25\u001b[0m\n\u001b[1;32m     14\u001b[0m prots_interrogated_by_screen \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39munique([\n\u001b[1;32m     15\u001b[0m     sub_entry\n\u001b[1;32m     16\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m entry \u001b[38;5;129;01min\u001b[39;00m VACV_screen_df[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUniProt_IDs\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m     17\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m sub_entry \u001b[38;5;129;01min\u001b[39;00m entry\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m;\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     18\u001b[0m ])\n\u001b[1;32m     20\u001b[0m coverage_list \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     21\u001b[0m     human_int_partner \u001b[38;5;129;01min\u001b[39;00m prots_interrogated_by_screen\n\u001b[1;32m     22\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m human_int_partner \u001b[38;5;129;01min\u001b[39;00m human_int_partners\n\u001b[1;32m     23\u001b[0m ]\n\u001b[0;32m---> 25\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mall\u001b[39m(coverage_list), (\n\u001b[1;32m     26\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNot all human interaction partners comprised in the PPI data set \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     27\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mare interrogated in the VACV screen!\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     28\u001b[0m )\n",
      "\u001b[0;31mAssertionError\u001b[0m: Not all human interaction partners comprised in the PPI data set are interrogated in the VACV screen!"
     ]
    }
   ],
   "source": [
    "# Now, extract the human interaction partners from the PPI pairs and\n",
    "# determine whether they are present in the VACV screen TSV file\n",
    "# The file with confirmed positive and reliable negative interactions\n",
    "# uses UniProt Accessions (in the VACV screen TSV file, the\n",
    "# corresponding column is incorrectly named \"UniProt_IDs\")\n",
    "import numpy as np\n",
    "\n",
    "human_int_partners = np.unique(VACV_WR_pos_and_neg_PPIs_df[\"Human_prot\"])\n",
    "\n",
    "# Bear in mind that the `UniProt_IDs` column also contains composite\n",
    "# entries, i.e. entries encompassing multiple UniProt Accessions\n",
    "# separated by semicolons in order to accommodate isoforms generated by\n",
    "# alternative splicing, etc.\n",
    "prots_interrogated_by_screen = np.unique([\n",
    "    sub_entry\n",
    "    for entry in VACV_screen_df[\"UniProt_IDs\"]\n",
    "    for sub_entry in entry.split(\";\")\n",
    "])\n",
    "\n",
    "coverage_list = [\n",
    "    human_int_partner in prots_interrogated_by_screen\n",
    "    for human_int_partner in human_int_partners\n",
    "]\n",
    "\n",
    "assert all(coverage_list), (\n",
    "    \"Not all human interaction partners comprised in the PPI data set \"\n",
    "    \"are interrogated in the VACV screen!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amount of human interaction partners exhibiting a UniProt accession mismatch: 63\n"
     ]
    }
   ],
   "source": [
    "n_human_int_partners_uniprot_accession_mismatch = (\n",
    "    len(coverage_list) - np.count_nonzero(coverage_list)\n",
    ")\n",
    "\n",
    "print(\n",
    "    \"Amount of human interaction partners exhibiting a UniProt \"\n",
    "    f\"accession mismatch: {n_human_int_partners_uniprot_accession_mismatch:,}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A0A075B749\n",
      "A0A0D9SG04\n",
      "A0A0J9YX62\n",
      "A0A0U1RRM6\n",
      "A0A1B0GTL5\n",
      "A0A2R8Y5A3\n",
      "A0A3B3IRW5\n",
      "A0A3B3IS91\n",
      "A0A3B3ISQ4\n",
      "A0N0N7\n",
      "A0N0Q3\n",
      "A3KPC7\n",
      "A4FTV9\n",
      "A6NFX8\n",
      "A6NNZ2\n",
      "A8ASI8\n",
      "A8MPP1\n",
      "A8MUS3\n",
      "B2R4P9\n",
      "B2R4R0\n",
      "B2R4S9\n",
      "B2RDW1\n",
      "B2ZZ89\n",
      "B4DJ51\n",
      "B4DLJ1\n",
      "B8ZZN6\n",
      "C9JQJ2\n",
      "D9YZV4\n",
      "D9ZGF2\n",
      "E5KTA5\n",
      "E9KL37\n",
      "E9PDI4\n",
      "F4ZW62\n",
      "F8VVA7\n",
      "F8VXC8\n",
      "F8VZQ9\n",
      "F8WBV6\n",
      "G3V5R9\n",
      "G5E9I4\n",
      "H3BSR6\n",
      "I0J062\n",
      "J3QK89\n",
      "K7EQ78\n",
      "K7ERV3\n",
      "P0DW81\n",
      "P22532\n",
      "P57053\n",
      "P61571\n",
      "P61572\n",
      "P61573\n",
      "P61574\n",
      "P61575\n",
      "P61576\n",
      "P61578\n",
      "P61579\n",
      "P86452\n",
      "Q0WX57\n",
      "Q548T7\n",
      "Q5VTE0\n",
      "Q6ZN40\n",
      "Q99878\n",
      "Q9UN81\n",
      "V9GZ56\n"
     ]
    }
   ],
   "source": [
    "# Output the UniProt Accessions of the proteins not interrogated in the\n",
    "# screen\n",
    "prots_with_uniprot_acc_mismatch = human_int_partners[\n",
    "    ~np.array(coverage_list)\n",
    "]\n",
    "\n",
    "for prot_with_mismatch in prots_with_uniprot_acc_mismatch:\n",
    "    print(prot_with_mismatch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A UniProt database query is performed for the UniProt accessions not\n",
    "# comprised in the VACV screen TSV file\n",
    "# The files are requested in XML format as this allows to extract the\n",
    "# official gene symbol in a systematic and reliable manner\n",
    "from biotite.database import uniprot\n",
    "\n",
    "# As no target directory to store files in is specified, the file\n",
    "# contents are stored in StringIO objects\n",
    "uniprot_not_queried = True\n",
    "\n",
    "while uniprot_not_queried:\n",
    "    try:\n",
    "        uniprot_io_objects = uniprot.fetch(\n",
    "            prots_with_uniprot_acc_mismatch,\n",
    "            format=\"xml\"\n",
    "        )\n",
    "        uniprot_not_queried = False\n",
    "    except ConnectionError:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate over the StringIO objects, access their contents via the\n",
    "# `.getvalue()` method and determine the official gene symbol\n",
    "# Determination of the official gene name is accomplished by means of\n",
    "# the `xml` library, which allows parsing XML files\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "off_gene_symbol_list = []\n",
    "\n",
    "for io_object, uniprot_accession in zip(\n",
    "    uniprot_io_objects, prots_with_uniprot_acc_mismatch\n",
    "):\n",
    "    xml_string = io_object.getvalue()\n",
    "    \n",
    "    root = ET.fromstring(xml_string)\n",
    "\n",
    "    # An XPath expression is used in order to find the official gene\n",
    "    # symbol\n",
    "    off_gene_symbol_entries = root.findall(\n",
    "        \"./{http://uniprot.org/uniprot}entry/\"\n",
    "        \"{http://uniprot.org/uniprot}gene/\"\n",
    "        \"{http://uniprot.org/uniprot}name[@type='primary']\"\n",
    "    )\n",
    "\n",
    "    # One protein can be encoded by multiple genes, which is why a\n",
    "    # sublist is generated for each protein containing the potentially\n",
    "    # multiple gene names\n",
    "    current_gene_name_list = []\n",
    "\n",
    "    for gene_symbol_entry in off_gene_symbol_entries:\n",
    "        current_gene_name_list.append(gene_symbol_entry.text)\n",
    "    \n",
    "    off_gene_symbol_list.append(current_gene_name_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0 1 7]\n"
     ]
    }
   ],
   "source": [
    "# Determine unique values for the amount of genes associated with a\n",
    "# UniProt Accession\n",
    "n_genes_per_prot_list = list(map(len, off_gene_symbol_list))\n",
    "\n",
    "print(np.unique(n_genes_per_prot_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amount of UniProt accessions without an associated gene: 1\n"
     ]
    }
   ],
   "source": [
    "# Determine the amount of UniProt accessions without an associated gene\n",
    "n_prots_without_gene = np.count_nonzero(\n",
    "    np.array(n_genes_per_prot_list) == 0\n",
    ")\n",
    "\n",
    "print(\n",
    "    \"Amount of UniProt accessions without an associated gene: \"\n",
    "    f\"{n_prots_without_gene}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\" encoding=\"UTF-8\"  standalone=\"no\" ?>\n",
      "<uniprot xmlns=\"http://uniprot.org/uniprot\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"http://uniprot.org/uniprot http://www.uniprot.org/docs/uniprot.xsd\">\n",
      "<entry dataset=\"TrEMBL\" created=\"2011-06-28\" modified=\"2025-02-05\" version=\"81\" xmlns=\"http://uniprot.org/uniprot\">\n",
      "  <accession>F4ZW62</accession>\n",
      "  <name>F4ZW62_HUMAN</name>\n",
      "  <protein>\n",
      "    <submittedName>\n",
      "      <fullName evidence=\"8\">NF45</fullName>\n",
      "    </submittedName>\n",
      "  </protein>\n",
      "  <organism evidence=\"8\">\n",
      "    <name type=\"scientific\">Homo sapiens</name>\n",
      "    <name type=\"common\">Human</name>\n",
      "    <dbReference type=\"NCBI Taxonomy\" id=\"9606\"/>\n",
      "    <lineage>\n",
      "      <taxon>Eukaryota</taxon>\n",
      "      <taxon>Metazoa</taxon>\n",
      "      <taxon>Chordata</taxon>\n",
      "      <taxon>Craniata</taxon>\n",
      "      <taxon>Vertebrata</taxon>\n",
      "      <taxon>Euteleostomi</taxon>\n",
      "      <taxon>Mammalia</taxon>\n",
      "      <taxon>Eutheria</taxon>\n",
      "      <taxon>Euarchontoglires</taxon>\n",
      "      <taxon>Primates</taxon>\n",
      "      <taxon>Haplorrhini</taxon>\n",
      "      <taxon>Catarrhini</taxon>\n",
      "      <taxon>Hominidae</taxon>\n",
      "      <taxon>Homo</taxon>\n",
      "    </lineage>\n",
      "  </organism>\n",
      "  <reference evidence=\"8\" key=\"1\">\n",
      "    <citation type=\"submission\" date=\"2010-09\" db=\"EMBL/GenBank/DDBJ databases\">\n",
      "      <title>Inhibition of virus replication by NF45 and NF90.</title>\n",
      "      <authorList>\n",
      "        <person name=\"Lin C.-C.\"/>\n",
      "      </authorList>\n",
      "    </citation>\n",
      "    <scope>NUCLEOTIDE SEQUENCE</scope>\n",
      "  </reference>\n",
      "  <comment type=\"interaction\">\n",
      "    <interactant intactId=\"EBI-10177680\">\n",
      "      <id>F4ZW62</id>\n",
      "    </interactant>\n",
      "    <interactant intactId=\"EBI-748961\">\n",
      "      <id>O95273</id>\n",
      "      <label>CCNDBP1</label>\n",
      "    </interactant>\n",
      "    <organismsDiffer>false</organismsDiffer>\n",
      "    <experiments>3</experiments>\n",
      "  </comment>\n",
      "  <comment type=\"interaction\">\n",
      "    <interactant intactId=\"EBI-10177680\">\n",
      "      <id>F4ZW62</id>\n",
      "    </interactant>\n",
      "    <interactant intactId=\"EBI-10177695\">\n",
      "      <id>P26641-2</id>\n",
      "      <label>EEF1G</label>\n",
      "    </interactant>\n",
      "    <organismsDiffer>false</organismsDiffer>\n",
      "    <experiments>3</experiments>\n",
      "  </comment>\n",
      "  <comment type=\"subcellular location\">\n",
      "    <subcellularLocation>\n",
      "      <location evidence=\"1\">Nucleus</location>\n",
      "    </subcellularLocation>\n",
      "  </comment>\n",
      "  <dbReference type=\"EMBL\" id=\"HQ322261\">\n",
      "    <property type=\"protein sequence ID\" value=\"AEE02016.1\"/>\n",
      "    <property type=\"molecule type\" value=\"mRNA\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"RefSeq\" id=\"NP_004506.2\">\n",
      "    <property type=\"nucleotide sequence ID\" value=\"NM_004515.3\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"AlphaFoldDB\" id=\"F4ZW62\"/>\n",
      "  <dbReference type=\"SMR\" id=\"F4ZW62\"/>\n",
      "  <dbReference type=\"IntAct\" id=\"F4ZW62\">\n",
      "    <property type=\"interactions\" value=\"3\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"Antibodypedia\" id=\"1451\">\n",
      "    <property type=\"antibodies\" value=\"492 antibodies from 39 providers\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"DNASU\" id=\"3608\"/>\n",
      "  <dbReference type=\"GeneID\" id=\"3608\"/>\n",
      "  <dbReference type=\"KEGG\" id=\"hsa:3608\"/>\n",
      "  <dbReference type=\"CTD\" id=\"3608\"/>\n",
      "  <dbReference type=\"DisGeNET\" id=\"3608\"/>\n",
      "  <dbReference type=\"VEuPathDB\" id=\"HostDB:ENSG00000143621\"/>\n",
      "  <dbReference type=\"OMA\" id=\"YLAIEMS\"/>\n",
      "  <dbReference type=\"OrthoDB\" id=\"5775647at2759\"/>\n",
      "  <dbReference type=\"PhylomeDB\" id=\"F4ZW62\"/>\n",
      "  <dbReference type=\"BioGRID-ORCS\" id=\"3608\">\n",
      "    <property type=\"hits\" value=\"658 hits in 1162 CRISPR screens\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"ExpressionAtlas\" id=\"F4ZW62\">\n",
      "    <property type=\"expression patterns\" value=\"baseline and differential\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"GO\" id=\"GO:0005829\">\n",
      "    <property type=\"term\" value=\"C:cytosol\"/>\n",
      "    <property type=\"evidence\" value=\"ECO:0000314\"/>\n",
      "    <property type=\"project\" value=\"HPA\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"GO\" id=\"GO:0005730\">\n",
      "    <property type=\"term\" value=\"C:nucleolus\"/>\n",
      "    <property type=\"evidence\" value=\"ECO:0007669\"/>\n",
      "    <property type=\"project\" value=\"Ensembl\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"GO\" id=\"GO:0005654\">\n",
      "    <property type=\"term\" value=\"C:nucleoplasm\"/>\n",
      "    <property type=\"evidence\" value=\"ECO:0000314\"/>\n",
      "    <property type=\"project\" value=\"HPA\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"GO\" id=\"GO:0003677\">\n",
      "    <property type=\"term\" value=\"F:DNA binding\"/>\n",
      "    <property type=\"evidence\" value=\"ECO:0007669\"/>\n",
      "    <property type=\"project\" value=\"UniProtKB-KW\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"FunFam\" id=\"1.10.1410.40:FF:000004\">\n",
      "    <property type=\"entry name\" value=\"Interleukin enhancer-binding factor 2\"/>\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"FunFam\" id=\"1.10.1410.40:FF:000010\">\n",
      "    <property type=\"entry name\" value=\"Interleukin enhancer-binding factor 2\"/>\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"FunFam\" id=\"3.30.460.10:FF:000093\">\n",
      "    <property type=\"entry name\" value=\"Interleukin enhancer-binding factor 2\"/>\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"Gene3D\" id=\"1.10.1410.40\">\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"Gene3D\" id=\"3.30.460.10\">\n",
      "    <property type=\"entry name\" value=\"Beta Polymerase, domain 2\"/>\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"InterPro\" id=\"IPR006561\">\n",
      "    <property type=\"entry name\" value=\"DZF_dom\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"InterPro\" id=\"IPR049402\">\n",
      "    <property type=\"entry name\" value=\"DZF_dom_C\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"InterPro\" id=\"IPR049401\">\n",
      "    <property type=\"entry name\" value=\"DZF_dom_N\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"InterPro\" id=\"IPR052134\">\n",
      "    <property type=\"entry name\" value=\"ILF2\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"InterPro\" id=\"IPR043519\">\n",
      "    <property type=\"entry name\" value=\"NT_sf\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"PANTHER\" id=\"PTHR46447\">\n",
      "    <property type=\"entry name\" value=\"INTERLEUKIN ENHANCER-BINDING FACTOR\"/>\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"PANTHER\" id=\"PTHR46447:SF1\">\n",
      "    <property type=\"entry name\" value=\"INTERLEUKIN ENHANCER-BINDING FACTOR 2\"/>\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"Pfam\" id=\"PF20965\">\n",
      "    <property type=\"entry name\" value=\"DZF_C\"/>\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"Pfam\" id=\"PF07528\">\n",
      "    <property type=\"entry name\" value=\"DZF_N\"/>\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"SMART\" id=\"SM00572\">\n",
      "    <property type=\"entry name\" value=\"DZF\"/>\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"SUPFAM\" id=\"SSF81301\">\n",
      "    <property type=\"entry name\" value=\"Nucleotidyltransferase\"/>\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"PROSITE\" id=\"PS50152\">\n",
      "    <property type=\"entry name\" value=\"25A_SYNTH_3\"/>\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <dbReference type=\"PROSITE\" id=\"PS51703\">\n",
      "    <property type=\"entry name\" value=\"DZF\"/>\n",
      "    <property type=\"match status\" value=\"1\"/>\n",
      "  </dbReference>\n",
      "  <proteinExistence type=\"evidence at protein level\"/>\n",
      "  <keyword evidence=\"4\" id=\"KW-0010\">Activator</keyword>\n",
      "  <keyword evidence=\"3\" id=\"KW-0238\">DNA-binding</keyword>\n",
      "  <keyword evidence=\"5\" id=\"KW-0804\">Transcription</keyword>\n",
      "  <keyword evidence=\"2\" id=\"KW-0805\">Transcription regulation</keyword>\n",
      "  <feature type=\"domain\" description=\"DZF\" evidence=\"7\">\n",
      "    <location>\n",
      "      <begin position=\"24\"/>\n",
      "      <end position=\"371\"/>\n",
      "    </location>\n",
      "  </feature>\n",
      "  <feature type=\"region of interest\" description=\"Disordered\" evidence=\"6\">\n",
      "    <location>\n",
      "      <begin position=\"1\"/>\n",
      "      <end position=\"20\"/>\n",
      "    </location>\n",
      "  </feature>\n",
      "  <feature type=\"region of interest\" description=\"Disordered\" evidence=\"6\">\n",
      "    <location>\n",
      "      <begin position=\"351\"/>\n",
      "      <end position=\"390\"/>\n",
      "    </location>\n",
      "  </feature>\n",
      "  <feature type=\"compositionally biased region\" description=\"Basic and acidic residues\" evidence=\"6\">\n",
      "    <location>\n",
      "      <begin position=\"356\"/>\n",
      "      <end position=\"367\"/>\n",
      "    </location>\n",
      "  </feature>\n",
      "  <feature type=\"compositionally biased region\" description=\"Acidic residues\" evidence=\"6\">\n",
      "    <location>\n",
      "      <begin position=\"368\"/>\n",
      "      <end position=\"390\"/>\n",
      "    </location>\n",
      "  </feature>\n",
      "  <evidence type=\"ECO:0000256\" key=\"1\">\n",
      "    <source>\n",
      "      <dbReference type=\"ARBA\" id=\"ARBA00004123\"/>\n",
      "    </source>\n",
      "  </evidence>\n",
      "  <evidence type=\"ECO:0000256\" key=\"2\">\n",
      "    <source>\n",
      "      <dbReference type=\"ARBA\" id=\"ARBA00023015\"/>\n",
      "    </source>\n",
      "  </evidence>\n",
      "  <evidence type=\"ECO:0000256\" key=\"3\">\n",
      "    <source>\n",
      "      <dbReference type=\"ARBA\" id=\"ARBA00023125\"/>\n",
      "    </source>\n",
      "  </evidence>\n",
      "  <evidence type=\"ECO:0000256\" key=\"4\">\n",
      "    <source>\n",
      "      <dbReference type=\"ARBA\" id=\"ARBA00023159\"/>\n",
      "    </source>\n",
      "  </evidence>\n",
      "  <evidence type=\"ECO:0000256\" key=\"5\">\n",
      "    <source>\n",
      "      <dbReference type=\"ARBA\" id=\"ARBA00023163\"/>\n",
      "    </source>\n",
      "  </evidence>\n",
      "  <evidence type=\"ECO:0000256\" key=\"6\">\n",
      "    <source>\n",
      "      <dbReference type=\"SAM\" id=\"MobiDB-lite\"/>\n",
      "    </source>\n",
      "  </evidence>\n",
      "  <evidence type=\"ECO:0000259\" key=\"7\">\n",
      "    <source>\n",
      "      <dbReference type=\"PROSITE\" id=\"PS51703\"/>\n",
      "    </source>\n",
      "  </evidence>\n",
      "  <evidence type=\"ECO:0000313\" key=\"8\">\n",
      "    <source>\n",
      "      <dbReference type=\"EMBL\" id=\"AEE02016.1\"/>\n",
      "    </source>\n",
      "  </evidence>\n",
      "  <sequence length=\"390\" mass=\"43062\" checksum=\"75BAD022DCD4EE01\" modified=\"2011-06-28\" version=\"1\">MRGDRGRGRGGRFGSRGGPGGGFRPFVPHIPFDFYLCEMAFPRVKPAPDETSFSEALLKRNQDLAPNSAEQASILSLVTKINNVIDNLIVAPGTFEVQIEEVRQVGSYKKGTMTTGHNVADLVVILKILPTLEAVAALGNKVVESLRAQDPSEVLTMLTNETGFEISSSDATVKILITTVPPNLRKLDPELHLDIKVLQSALAAIRHARWFEENASQSTVKVLIRLLKDLRIRFPGFEPLTPWILDLLGHYAVMNNPTRQPLALNVAYRRCLQILAAGLFLPGSVGITDPCESGNFRVHTVMTLEQQDMVCYTAQTLVRILSHGGFRKILGQEGDASYLASEISTWDGVIVTPSEKAYEKPPEKKEGEEEEENTEEPPQGEEEESMETQE</sequence>\n",
      "</entry>\n",
      "<copyright>\n",
      "Copyrighted by the UniProt Consortium, see https://www.uniprot.org/terms Distributed under the Creative Commons Attribution (CC BY 4.0) License\n",
      "</copyright>\n",
      "</uniprot>\n"
     ]
    }
   ],
   "source": [
    "# Determine the protein's identity and subject its XML file to closer\n",
    "# scrutiny\n",
    "orphan_prot_idx = n_genes_per_prot_list.index(0)\n",
    "\n",
    "orphan_xml_str = uniprot_io_objects[orphan_prot_idx].getvalue()\n",
    "\n",
    "print(orphan_xml_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Strangely enough, while the XML file does specify an NCBI gene ID, it\n",
    "# does not indicate the corresponding gene name\n",
    "# Thus, for this UniProt accession, the official gene symbol is manually\n",
    "# added to the correct sublist\n",
    "off_gene_symbol_list[orphan_prot_idx].append(\"ILF2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UniProt accession A0A3B3IS91's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession A3KPC7's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession A4FTV9's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession A6NNZ2's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession A8MPP1's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession B2R4P9's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession B2R4R0's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession B2R4S9's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession B4DJ51's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession I0J062's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession P0DW81's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession P22532's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession P57053's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession P61571's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession P61572's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession P61573's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession P61574's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession P61575's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession P61576's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession P61578's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession P61579's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession P86452's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession Q0WX57's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession Q5VTE0's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession Q99878's gene name(s) do(es) not match any of the gene names in the VACV screen!\n",
      "UniProt accession Q9UN81's gene name(s) do(es) not match any of the gene names in the VACV screen!\n"
     ]
    }
   ],
   "source": [
    "# Now that each and every UniProt accession has at least one associated\n",
    "# gene symbol, it is investigated for each UniProt accession whether at\n",
    "# least one of them is interrogated in the VACV screen\n",
    "gene_symbols_in_screen = np.unique(\n",
    "    VACV_screen_df[\"Name\"]\n",
    ")\n",
    "\n",
    "acc_mismatch_subset_with_gene_name_mismatch = []\n",
    "\n",
    "for prot, sub_list in zip(prots_with_uniprot_acc_mismatch, off_gene_symbol_list):\n",
    "    presence_list = [\n",
    "        gene_symbol in gene_symbols_in_screen\n",
    "        for gene_symbol in sub_list\n",
    "    ]\n",
    "\n",
    "    if not any(presence_list):\n",
    "        print(\n",
    "            f\"UniProt accession {prot}'s gene name(s) do(es) not match \"\n",
    "            \"any of the gene names in the VACV screen!\"\n",
    "        )\n",
    "        acc_mismatch_subset_with_gene_name_mismatch.append(prot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The precise amount of proteins not matching both the UniProt accessions\n",
      "and the gene names in the VACV screen is 26.\n"
     ]
    }
   ],
   "source": [
    "n_prots_acc_and_name_mismatch = len(\n",
    "    acc_mismatch_subset_with_gene_name_mismatch\n",
    ")\n",
    "\n",
    "print(\n",
    "    \"The precise amount of proteins not matching both the UniProt \"\n",
    "    \"accessions\\nand the gene names in the VACV screen is \"\n",
    "    f\"{n_prots_acc_and_name_mismatch:,}.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique numbers of associated gene IDs: [ 0  1  2  3  5  7 14]\n"
     ]
    }
   ],
   "source": [
    "# Upon closer scrutiny of the proteins with both UniProt accession and\n",
    "# gene name mismatch, it emerges that although the corresponding gene\n",
    "# indeed is present in the screen, a gene name different from the one\n",
    "# used by NCBI is specified\n",
    "# Therefore, for those proteins the name of which does not match the one\n",
    "# used NCBI, it is investigated whether the corresponding gene is\n",
    "# covered by the screen\n",
    "# This is accomplished by means of the NCBI gene ID\n",
    "# Performing this coverage investigation in multiple steps is necessary\n",
    "# as e.g. not all proteins the gene symbol of which matches the one used\n",
    "# by NCBI have an associated NCBI gene ID in their XML file\n",
    "\n",
    "# As the different UniProt entries have different amounts of associated\n",
    "# genes, the unique numbers of associated gene names are determined\n",
    "n_associated_gene_IDs = []\n",
    "\n",
    "for io_object, uniprot_accession in zip(\n",
    "    uniprot_io_objects, prots_with_uniprot_acc_mismatch\n",
    "):\n",
    "    if uniprot_accession in acc_mismatch_subset_with_gene_name_mismatch:\n",
    "        xml_string = io_object.getvalue()\n",
    "\n",
    "        root = ET.fromstring(xml_string)\n",
    "\n",
    "        # An XPath expression is used in order to find the NCBI gene ID\n",
    "        gene_id_entries = root.findall(\n",
    "            \"./{http://uniprot.org/uniprot}entry/\"\n",
    "            \"{http://uniprot.org/uniprot}dbReference[@type='GeneID']\"\n",
    "        )\n",
    "\n",
    "        n_associated_gene_IDs.append(len(gene_id_entries))\n",
    "\n",
    "print(\n",
    "    \"Unique numbers of associated gene IDs: \"\n",
    "    f\"{np.unique(n_associated_gene_IDs)}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UniProt accessions of proteins with no associated gene ID:\n",
      "A0A3B3IS91\n",
      "A6NNZ2\n",
      "A8MPP1\n",
      "P0DW81\n",
      "P61571\n",
      "P61572\n",
      "P61573\n",
      "P61574\n",
      "P61575\n",
      "P61576\n",
      "P61578\n",
      "P61579\n",
      "Q5VTE0\n",
      "Q9UN81\n",
      "\n",
      "Uniprot accessions of proteins with one associated gene ID:\n",
      "A3KPC7 -> 85235\n",
      "I0J062 -> 101927423\n",
      "P22532 -> 6703\n",
      "P57053 -> 102724334\n",
      "P86452 -> 100381270\n",
      "Q99878 -> 8331\n",
      "\n",
      "UniProt accessions of proteins with multiple associated gene IDs:\n",
      "A4FTV9 -> ['8329', '8330', '8332', '8336', '8969']\n",
      "B2R4P9 -> ['3020', '3021']\n",
      "B2R4R0 -> ['121504', '554313', '8294', '8359', '8360', '8361', '8362', '8363', '8364', '8365', '8366', '8367', '8368', '8370']\n",
      "B2R4S9 -> ['8339', '8343', '8344', '8346', '8347']\n",
      "B4DJ51 -> ['801', '805', '808']\n",
      "Q0WX57 -> ['728369', '728373', '728379', '728393', '728400', '728405', '728419']\n"
     ]
    }
   ],
   "source": [
    "# Based on the amount of associated gene IDs, the UniProt entries\n",
    "# exhibiting both a UniProt accession and a gene name mismatch are\n",
    "# divided into three groups\n",
    "# The first group encompasses proteins without any associated gene IDs\n",
    "# The second group comprises proteins with only one associated gene ID\n",
    "# The third group contains proteins with more than one associated gene\n",
    "# ID\n",
    "\n",
    "# Output the UniProt accessions for each of the three groups\n",
    "uniprot_accs_group_one = []\n",
    "\n",
    "uniprot_accs_group_two = []\n",
    "gene_ids_group_two = []\n",
    "\n",
    "uniprot_accs_group_three = []\n",
    "gene_ids_group_three = []\n",
    "\n",
    "for io_object, uniprot_accession in zip(\n",
    "    uniprot_io_objects, prots_with_uniprot_acc_mismatch\n",
    "):\n",
    "    if uniprot_accession in acc_mismatch_subset_with_gene_name_mismatch:\n",
    "        xml_string = io_object.getvalue()\n",
    "\n",
    "        root = ET.fromstring(xml_string)\n",
    "\n",
    "        # An XPath expression is used in order to find the NCBI gene ID\n",
    "        gene_id_entries = root.findall(\n",
    "            \"./{http://uniprot.org/uniprot}entry/\"\n",
    "            \"{http://uniprot.org/uniprot}dbReference[@type='GeneID']\"\n",
    "        )\n",
    "\n",
    "        n_gene_ids = len(gene_id_entries)\n",
    "\n",
    "        if n_gene_ids == 0:\n",
    "            uniprot_accs_group_one.append(uniprot_accession)\n",
    "        elif n_gene_ids == 1:\n",
    "            uniprot_accs_group_two.append(uniprot_accession)\n",
    "            gene_id = gene_id_entries[0].attrib[\"id\"]\n",
    "            gene_ids_group_two.append(gene_id)\n",
    "        else:\n",
    "            uniprot_accs_group_three.append(uniprot_accession)\n",
    "            gene_ids = [\n",
    "                gene_id_entry.attrib[\"id\"]\n",
    "                for gene_id_entry in gene_id_entries\n",
    "            ]\n",
    "            gene_ids_group_three.append(gene_ids)\n",
    "\n",
    "print(\"UniProt accessions of proteins with no associated gene ID:\")\n",
    "for accession in uniprot_accs_group_one:\n",
    "    print(accession)\n",
    "print()\n",
    "\n",
    "print(\"Uniprot accessions of proteins with one associated gene ID:\")\n",
    "for accession, gene_id in zip(uniprot_accs_group_two, gene_ids_group_two):\n",
    "    print(f\"{accession} -> {gene_id}\")\n",
    "print()\n",
    "\n",
    "print(\"UniProt accessions of proteins with multiple associated gene IDs:\")\n",
    "for accession, gene_ids in zip(uniprot_accs_group_three, gene_ids_group_three):\n",
    "    print(f\"{accession} -> {gene_ids}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The UniProt IDs of each of the three groups are subjected to closer\n",
    "# scrutiny\n",
    "\n",
    "# With regard to the first group (no associated gene ID), it emerges\n",
    "# that the gene IDs can be found out manually by querying the NCBI\n",
    "# database\n",
    "# While some genes are interrogated by the VACV screen, others are not\n",
    "# A0A3B3IS91 -> 5428, present\n",
    "# A6NNZ2 -> 260334 (TUBB8B), not present\n",
    "# A8MPP1 -> 100302090 (DDX11L8), not present\n",
    "# P0DW81 -> 101929977 (MARCHF6-DT), not present\n",
    "# P61571 -> 100616103 (ERVK-21), not present\n",
    "# P61572 -> 105376906 (ERVK-19), not present\n",
    "# P61573 -> 100616102 (ERVK-9), not present\n",
    "# P61574 -> 124902738 (LOC124902738), not present\n",
    "# P61575 -> 619465 (ERVK-8), not present\n",
    "# P61576 -> no gene available\n",
    "# P61578 -> 100616411 (ERVK-16), not present\n",
    "# P61579 -> 100862683 (ERVK-25), not presen\n",
    "# Q5VTE0 -> 158078 (EEF1A1P5), not present\n",
    "# Q9UN81 -> no gene available\n",
    "\n",
    "# Investigating the UniProt accessions of the second group (one\n",
    "# associated gene ID) reveals the following:\n",
    "# A3KPC7 -> 85235, present\n",
    "# I0J062 -> 101927423, not present\n",
    "# P22532 -> 6703, not present\n",
    "# P57053 -> 102724334, not present\n",
    "# P86452 -> 100381270, not present\n",
    "# Q99878 -> 8331, not present\n",
    "\n",
    "# As to the third group (multiple associated gene IDs), many UniProt\n",
    "# accessions are covered by multiple gene IDs\n",
    "#\n",
    "# A4FTV9 (Histone H2A) -> ['8329', '8330', '8332', '8336', '8969']; gene\n",
    "# ids 8330, 8332, 8336 as well as 8969 are present\n",
    "#\n",
    "# B2R4P9 (Histone H3) -> ['3020', '3021']; gene ids 3020 as well as 3021\n",
    "# are present\n",
    "#\n",
    "# B2R4R0 (Histone H4) -> ['121504', '554313', '8294', '8359', '8360',\n",
    "# '8361', '8362', '8363', '8364', '8365', '8366', '8367', '8368',\n",
    "# '8370']; gene ids 121504, 554313, 8359, 8360, 8361, 8362, 8363, 8364,\n",
    "# 8365, 8366, 8367, 8368 as well as 8370 are present\n",
    "#\n",
    "# B2R4S9 (Histone H2B) -> ['8339', '8343', '8344', '8346', '8347']; gene\n",
    "# ids 8339, 8343, 8344, 8346 as well as 8347 are present\n",
    "#\n",
    "# B4DJ51 (Calmodulin 1) -> ['801', '805', '808']; gene ids 801, 805 as\n",
    "# well as 808 are present\n",
    "#\n",
    "# Q0WX57 (Ubiquitin carboxyl-terminal hydrolase 17-like protein 24) ->\n",
    "# ['728369', '728373', '728379', '728393', '728400', '728405', '728419'];\n",
    "# no gene ids are present\n",
    "\n",
    "# The question arises how to treat the multiple gene ids covering one\n",
    "# and the same UniProt accession\n",
    "# Maybe just take the highest intensity? -> Ask Artur!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now that the coverage of the human interaction partners of the\n",
    "# human-VACV PPIs has been exhaustively analysed, a dictionary is\n",
    "# created mapping each UniProt accession of a human interaction partner\n",
    "# to the corresponding NCBI gene ID\n",
    "# Just as the analysis, the dictionary creation is accomplished in a\n",
    "# stepwise manner\n",
    "uniprot_acc_gene_id_dict = {}\n",
    "\n",
    "# Start with the human interaction partners the gene name of which\n",
    "# matches the official gene symbols listed in the screen TSV file\n",
    "for int_partner in human_int_partners:\n",
    "    if int_partner in prots_interrogated_by_screen:\n",
    "        # As the `UniProt_IDs` column may contain multiple UniProt\n",
    "        # accessions separated from each other via semicolons, it is\n",
    "        # checked for the presence of substrings rather than for exact\n",
    "        # matches\n",
    "        # Also bear in mind that selecting a column yields a Pandas\n",
    "        # Series\n",
    "        # It is likely that there are multiple rows for the interaction\n",
    "        # partner at hand; however, as all rows are supposed to contain\n",
    "        # the same gene ID, simply the first gene ID is selected\n",
    "        current_gene_id = VACV_screen_df.loc[\n",
    "            VACV_screen_df[\"UniProt_IDs\"].str.contains(int_partner),\n",
    "            \"ID\"\n",
    "        ].iloc[0]\n",
    "\n",
    "        # The rationale behind setting a list as value instead of the\n",
    "        # mere gene ID is that there are proteins with multiple\n",
    "        # associated gene IDs\n",
    "        # Consistently setting a list as value simplifies the\n",
    "        # implementation of downstream code\n",
    "        uniprot_acc_gene_id_dict[int_partner] = [current_gene_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "737\n"
     ]
    }
   ],
   "source": [
    "print(len(uniprot_acc_gene_id_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, address the human interaction partners exhibiting a UniProt\n",
    "# accession mismatch\n",
    "for prot, sub_list in zip(prots_with_uniprot_acc_mismatch, off_gene_symbol_list):\n",
    "    presence_list = [\n",
    "        gene_symbol in gene_symbols_in_screen\n",
    "        for gene_symbol in sub_list\n",
    "    ]\n",
    "\n",
    "    if any(presence_list):\n",
    "        # Determine the precise gene symbol present in the screen; it is\n",
    "        # always just one\n",
    "        present_symbol = np.array(sub_list)[np.array(presence_list)][0]\n",
    "        \n",
    "        current_gene_id = VACV_screen_df.loc[\n",
    "            VACV_screen_df[\"Name\"] == present_symbol,\n",
    "            \"ID\"\n",
    "        ].iloc[0]\n",
    "\n",
    "        uniprot_acc_gene_id_dict[prot] = [current_gene_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "774\n"
     ]
    }
   ],
   "source": [
    "print(len(uniprot_acc_gene_id_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# As a next step, turn to human interaction partners exhibiting both a\n",
    "# UniProt accession mismatch and a gene name mismatch\n",
    "# For these human interaction partners, three subgroups could be\n",
    "# identified, namely human interaction partners with no, one and more\n",
    "# than one associated gene IDs\n",
    "\n",
    "# For proteins with one and multiple associated gene IDs, the respective\n",
    "# gene IDs are set as value in the dictionary\n",
    "for uniprot_accession, gene_id in zip(\n",
    "    uniprot_accs_group_two, gene_ids_group_two\n",
    "):\n",
    "    uniprot_acc_gene_id_dict[uniprot_accession] = [gene_id]\n",
    "\n",
    "for uniprot_accession, gene_id_list in zip(\n",
    "    uniprot_accs_group_three, gene_ids_group_three\n",
    "):\n",
    "    uniprot_acc_gene_id_dict[uniprot_accession] = gene_id_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "786\n"
     ]
    }
   ],
   "source": [
    "print(len(uniprot_acc_gene_id_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finally, deal with proteins with no associated gene ID\n",
    "# The gene IDs for these proteins were found out by manually querying\n",
    "# the NCBI database\n",
    "# Thus, a list of gene IDs has to be generated manually as well\n",
    "gene_ids_group_one = [\n",
    "    5428, 260334, 100302090, 101929977, 100616103, 105376906, 100616102,\n",
    "    124902738, 619465, None, 100616411, 100862683, 158078, None\n",
    "]\n",
    "\n",
    "for uniprot_accession, gene_id in zip(\n",
    "    uniprot_accs_group_one, gene_ids_group_one\n",
    "):\n",
    "    uniprot_acc_gene_id_dict[uniprot_accession] = [gene_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "800\n"
     ]
    }
   ],
   "source": [
    "print(len(uniprot_acc_gene_id_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform a couple of final sanity checks\n",
    "# One of them consists of verifying that the total amount of UniProt\n",
    "# accessions in the dictionary equals the total amount of unique UniProt\n",
    "# accessions occurring the combined VACV data set\n",
    "assert len(uniprot_acc_gene_id_dict) == len(human_int_partners), (\n",
    "    \"Not all unique UniProt accessions occurring in the combined VACV \"\n",
    "    \"data set have been added to the dictionary!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Another sanity check involves verifying that all dictionary values are\n",
    "# of data type `list`\n",
    "dict_vals = [value for value in uniprot_acc_gene_id_dict.values()]\n",
    "\n",
    "assert all([isinstance(value, list) for value in dict_vals]), (\n",
    "    \"Not all dictionary values are a list!\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now that all sanity checks have successfully been passed, the\n",
    "# dictionary mapping UniProt accessions to NCBI gene IDs is pickled,\n",
    "# i.e. saved to a file\n",
    "import pickle\n",
    "\n",
    "# Bear in mind that in the context of working with files, the `with`\n",
    "# context manager is preferred as it automatically takes care of closing\n",
    "# files, even in case of errors/exceptions\n",
    "with open(\n",
    "    \"dictionary_mapping_UniProt_accessions_of_combined_VACV_WR_data_\"\\\n",
    "    \"set_to_gene_IDs.pkl\",\n",
    "    \"wb\"\n",
    ") as f:\n",
    "    pickle.dump(uniprot_acc_gene_id_dict, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinformatics",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
