{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "166b6ba4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nThe purpose of this Jupyter notebook is to determine the optimal\\nthresholds from ROC curves in order to suggest novel potential\\ntherapeutic targets for VACV infection.\\n\\nTo be more precise, median-based intensity refinement turned out to\\nperform best, which is why its ROC curves are used to this end. One\\nthreshold is chosen for early and late intensities each.\\n\\nSubsequent to performing median-based intensity refinement and min-max\\nnormalization on the entire screen, the optimal thresholds are applied\\nto identify hits. In detail, the two thresholds are first applied\\nseparately on all genes/proteins, yielding two separate lists of hits.\\nProteins with a min-max normalized value greater than or equal to the\\nthreshold value are considered hits. In a subsequent step, the two hit\\nlists are merged, i.e. the union is formed.\\n\\nThe optimal threshold is determined using Youden's J index, which is\\ndefined as the difference between TPR and FPR (i.e. TPR - FPR).\\nSpecifically, the optimal threshold is the one associated with the\\nmaximum value of Youden's J index.\\n\""
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "The purpose of this Jupyter notebook is to determine the optimal\n",
    "thresholds from ROC curves in order to suggest novel potential\n",
    "therapeutic targets for VACV infection.\n",
    "\n",
    "To be more precise, median-based intensity refinement turned out to\n",
    "perform best, which is why its ROC curves are used to this end. One\n",
    "threshold is chosen for early and late intensities each.\n",
    "\n",
    "Subsequent to performing median-based intensity refinement and min-max\n",
    "normalization on the entire screen, the optimal thresholds are applied\n",
    "to identify hits. In detail, the two thresholds are first applied\n",
    "separately on all genes/proteins, yielding two separate lists of hits.\n",
    "Proteins with a min-max normalized value greater than or equal to the\n",
    "threshold value are considered hits. In a subsequent step, the two hit\n",
    "lists are merged, i.e. the union is formed.\n",
    "\n",
    "The optimal threshold is determined using Youden's J index, which is\n",
    "defined as the difference between TPR and FPR (i.e. TPR - FPR).\n",
    "Specifically, the optimal threshold is the one associated with the\n",
    "maximum value of Youden's J index.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3f36d5f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dbb30e29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the two TSV files containing FPR, TPR and threshold value triples\n",
    "path_to_early_tsv = (\n",
    "    \"/Users/jacobanter/Documents/Code/VACV_screen/Processing_Dharmacon_\"\n",
    "    \"pooled_genome_1_and_2_subset/roc_curves_and_related_data/refined_\"\n",
    "    \"normalized_intensity_values/median_refinement/roc_curve_data_\"\n",
    "    \"early_VoronoiCells.tsv\"\n",
    ")\n",
    "\n",
    "early_roc_curve_df = pd.read_csv(\n",
    "    path_to_early_tsv,\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "\n",
    "path_to_late_tsv = (\n",
    "    \"/Users/jacobanter/Documents/Code/VACV_screen/Processing_Dharmacon_\"\n",
    "    \"pooled_genome_1_and_2_subset/roc_curves_and_related_data/refined_\"\n",
    "    \"normalized_intensity_values/median_refinement/roc_curve_data_late_\"\n",
    "    \"VoronoiCells.tsv\"\n",
    ")\n",
    "\n",
    "late_roc_curve_df = pd.read_csv(\n",
    "    path_to_late_tsv,\n",
    "    sep=\"\\t\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "25260481",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Introduce a new column storing the difference between TPR and FPR for\n",
    "# each row\n",
    "early_roc_curve_df[\"J_index\"] = (\n",
    "    early_roc_curve_df[\"TPR\"]\n",
    "    -\n",
    "    early_roc_curve_df[\"FPR\"]\n",
    ")\n",
    "\n",
    "late_roc_curve_df[\"J_index\"] = (\n",
    "    late_roc_curve_df[\"TPR\"]\n",
    "    -\n",
    "    late_roc_curve_df[\"FPR\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "896ba380",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Overwrite the TSV files with the extended DataFrames\n",
    "early_roc_curve_df.to_csv(\n",
    "    path_to_early_tsv,\n",
    "    sep=\"\\t\",\n",
    "    index=False\n",
    ")\n",
    "\n",
    "late_roc_curve_df.to_csv(\n",
    "    path_to_late_tsv,\n",
    "    sep=\"\\t\",\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8b35f979",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Optimal threshold for early intensities: 0.404\n",
      "Optimal threshold for late intensities: 0.286\n"
     ]
    }
   ],
   "source": [
    "# Now, for early and late intensities individually, determine the\n",
    "# threshold associated with the maximum J index\n",
    "early_best_threshold = early_roc_curve_df.loc[\n",
    "    early_roc_curve_df[\"J_index\"].idxmax(), \"Threshold\"\n",
    "]\n",
    "\n",
    "late_best_threshold = late_roc_curve_df.loc[\n",
    "    late_roc_curve_df[\"J_index\"].idxmax(), \"Threshold\"\n",
    "]\n",
    "\n",
    "print(\n",
    "    f\"Optimal threshold for early intensities: {early_best_threshold:.3f}\"\n",
    ")\n",
    "print(\n",
    "    f\"Optimal threshold for late intensities: {late_best_threshold:.3f}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0fa5428e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the optimal thresholds to a Pandas DataFrame\n",
    "optimal_thresholds_df = pd.DataFrame(\n",
    "    data={\n",
    "        \"threshold_type\": [\"early\", \"late\"],\n",
    "        \"threshold_value\": [early_best_threshold, late_best_threshold]\n",
    "    }\n",
    ")\n",
    "\n",
    "optimal_thresholds_df.to_csv(\n",
    "    \"/Users/jacobanter/Documents/Code/VACV_screen/Processing_Dharmacon_\"\n",
    "    \"pooled_genome_1_and_2_subset/roc_curves_and_related_data/refined_\"\n",
    "    \"normalized_intensity_values/median_refinement/optimal_thresholds.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "e1f4bdf2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.4036220452743436\n",
      "0.2860976796049103\n"
     ]
    }
   ],
   "source": [
    "# As a test, load the TSV file storing the optimal thresholds and try to\n",
    "# read them\n",
    "optimal_thresholds_df = pd.read_csv(\n",
    "    \"/Users/jacobanter/Documents/Code/VACV_screen/Processing_Dharmacon_\"\n",
    "    \"pooled_genome_1_and_2_subset/roc_curves_and_related_data/refined_\"\n",
    "    \"normalized_intensity_values/median_refinement/optimal_thresholds.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=\"threshold_type\"\n",
    ")\n",
    "\n",
    "print(\n",
    "    optimal_thresholds_df.loc[\"early\", \"threshold_value\"]\n",
    ")\n",
    "\n",
    "print(\n",
    "    optimal_thresholds_df.loc[\"late\", \"threshold_value\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27452c46",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinformatics",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
